---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
# get a fire peak data table called `df`
source("Rscripts/utils.R")
getmode <- function(v) {
   uniqv <- unique(v)
   uniqv[which.max(tabulate(match(v, uniqv)))]
}
library(rlang)
#library(gmm)
# Load the package
library(mclust)
```

```{r}
all_covs = fread("results/PS00357_COLO829T_1/FDR-peaks/FDR.track.bed.gz") %>%
    # require 95% phasing
    filter(
        `#chrom` != "chrX",
        `#chrom` != "chrY",
    ) %>%
    filter(coverage * 0.95 <= coverage_H1 + coverage_H2) %>%
    filter(
        pmin(coverage_H1, coverage_H2) > 1,
    ) %>%
    sample_n(1e6) %>%
    data.table
model_cov = all_covs %>% filter(
    coverage < 200,
    coverage > 30
)

pdf("tmp.pdf")
hist(model_cov$coverage_H1, breaks=100, xlim=c(0,100))
hist(model_cov$coverage_H2, breaks=100, col="red", xlim=c(0,100))
hist(model_cov$coverage, breaks=100, col="blue", xlim=c(0,100))
hist(model_cov$coverage_H1 + model_cov$coverage_H2, breaks=100, col="blue", xlim=c(0,100))
dev.off()
nrow(fire)

m = Mclust(c(model_cov$coverage_H1, model_cov$coverage_H2), G=seq(2))
m$parameters$mean
```

```{r}
# Tumor PS00357
fire = fread("results/PS00357_COLO829T_1/FDR-peaks/FDR-FIRE-peaks.bed.gz") %>%
    # require 95% phasing
    filter(
        `#chrom` != "chrX",
        `#chrom` != "chrY",
    ) %>%
    filter(coverage * 0.9 <= coverage_H1 + coverage_H2) %>%
    filter(
        pmin(coverage_H1, coverage_H2) > 1,
    ) %>%
    mutate(
        high_cov_hap = case_when(
            coverage_H1 >= coverage_H2 ~ coverage_H1,
            TRUE ~ coverage_H2
        ),
        high_cov_hap_fire = case_when(
            fire_coverage_H1 >= fire_coverage_H2 ~ fire_coverage_H1,
            TRUE ~ fire_coverage_H2
        ),
        low_cov_hap = case_when(
            coverage_H1 < coverage_H2 ~ coverage_H1,
            TRUE ~ coverage_H2
        ),
        low_cov_hap_fire = case_when(
            fire_coverage_H1 < fire_coverage_H2 ~ fire_coverage_H1,
            TRUE ~ fire_coverage_H2
        ),
    )
sd_range = 3
mean_hap_cov = mean(fire$coverage_H1 + fire$coverage_H2)/2
lower_hap_cov_limit = mean_hap_cov - sd_range * sqrt(mean_hap_cov)
upper_hap_cov_limit = mean_hap_cov + sd_range * sqrt(mean_hap_cov)
print(glue("Mean coverage: {mean_hap_cov}\tLower limit: {lower_hap_cov_limit}\tUpper limit: {upper_hap_cov_limit}"))

norm_vs_dup_df_wide = fire %>% 
    mutate(
        normal_low_cov = (low_cov_hap > lower_hap_cov_limit), #& (low_cov_hap < upper_hap_cov_limit),
        is_dup = (high_cov_hap >= 2*low_cov_hap),
    ) %>%
    mutate(
        percent_accessible_dup = 100*high_cov_hap_fire / high_cov_hap,
        percent_accessible_norm = 100*low_cov_hap_fire / low_cov_hap,
    ) %>%
    mutate(
        norm_minus_dup = percent_accessible_norm - percent_accessible_dup 
    )
    
mean(norm_vs_dup_df_wide$normal_low_cov)
mean(norm_vs_dup_df_wide$is_dup)    
    
norm_vs_dup_df = norm_vs_dup_df_wide %>%
    pivot_longer(
        cols = c("percent_accessible_dup", "percent_accessible_norm"),
        names_to = "type",
        values_to = "percent"
    )

norm_vs_dup_df_coverage = norm_vs_dup_df_wide %>%
    pivot_longer(
        cols = c("high_cov_hap", "low_cov_hap"),
        names_to = "type",
        values_to = "coverage_hap"
    )

table(norm_vs_dup_df_wide[,c("is_dup", "normal_low_cov")])
```

```{r}
```

```{r}
norm_vs_dup_df_wide %>%
    rename(
        `lower coverage haplotype` = low_cov_hap,
        `higher coverage haplotype` = high_cov_hap,
        `haplotype 1` = coverage_H1,
        `haplotype 2` = coverage_H2,
    ) %>%
    pivot_longer(
        cols = c("lower coverage haplotype", "higher coverage haplotype", "haplotype 1", "haplotype 2"),
        names_to = "type",
        values_to = "coverages"
    ) %>%
    select(type, coverages, is_dup, normal_low_cov) %>%
    mutate(
        per = rank(coverages)/n(),
    ) %>%
    filter(per < 0.99) %>%
    ggplot(
        aes(
            x=coverages,
            fill=is_dup, #& normal_low_cov,
        )
    ) +
    scale_fill_manual(
        "Two fold coverage difference\nbetween H1 and H2",
        values=c("skyblue", "darkred"),
    ) +
    geom_histogram(binwidth=1, position="stack", alpha=0.7) +
    scale_x_continuous("Coverage") +
    facet_col(~type) +
    my_grid() +
    theme(legend.position="top")
my_ggsave("Figures/COLO-available-template-total-coverages.pdf", height=3, width=3) 
```

```{r}
```

```{r}
norm_vs_dup_df %>%
    filter(is_dup, normal_low_cov) %>%
    ggplot(aes(fill=type, x=percent)) +
    #geom_histogram(binwidth=2, position="identity", alpha=0.7) +
    geom_histogram(
        aes(y=100*after_stat(density)),
        binwidth=5,
        alpha=0.7,
        position="dodge",
    ) +
    scale_fill_manual("",
        labels=rev(c("Duplicated haplotype", "Normal haplotype")),
        values=rev(c("darkred", "skyblue")),
    ) +
    scale_x_continuous("Percent accessible", limits=c(0, 100)) +
    my_grid() +
    theme(legend.position="top")
my_ggsave("Figures/COLO-available-template.pdf", height=3, width=3) 

norm_vs_dup_df_wide %>%
    filter(normal_low_cov, is_dup) %>%
    ggplot(
        aes(
            x=percent_accessible_norm,
            y=percent_accessible_dup,
        )
    ) + 
    geom_hex(bins=30) + scale_fill_distiller("", palette = "Spectral", trans="log10") +
    #geom_smooth(method="lm", se=FALSE, color="black", size=0.5) +
    geom_abline(slope=1, intercept=0, color="black", linetype="dashed") +
    stat_cor()+
    scale_x_continuous("Percent accessible (low coverage haplotype)", limits=c(0,100)) +
    scale_y_continuous("Percent accessible (high coverage haplotype)", limits=c(0,100)) +
    coord_fixed() +
    #facet_col(~is_dup) + 
    my_grid()
my_ggsave("Figures/COLO-available-template-scatter.pdf", height=3, width=3) 


norm_vs_dup_df_wide %>%
    filter(normal_low_cov) %>%
    ggplot(aes(x=norm_minus_dup, fill=is_dup)) + 
    geom_density(alpha=0.4,
        #aes(y = after_stat(scaled))
    ) +
    scale_x_continuous("Percent accessibility difference\n(low coverage haplotype - high coverage haplotype)", limits=c(-100,100)) +
    theme_minimal_grid() +
    #facet_col(~is_dup)+
    scale_fill_manual("",
        labels=c("Regions with normal coverage", "Regions with one duplicated haplotype"),
        values=c("skyblue", "darkred"),
    ) +
    theme(legend.position="top")
my_ggsave("Figures/COLO-available-template-diff.pdf")


# plot the coverage of the duplicated haplotype vs the normal haplotype
norm_vs_dup_df_coverage %>%
    mutate( 
        is_dup = case_when(
            is_dup ~ "One haplotype has >2 fold more coverage",
            TRUE ~ "Both haplotypes have similar coverage",
        )
    ) %>%
    #filter(is_dup) %>%
    ggplot(aes(fill=type, x=coverage_hap)) +
    geom_density(
        #aes(y = after_stat(scaled)),
        alpha=0.75, adjust=.6
    ) +
    scale_fill_manual("",
        labels=(c("higher coverage haplotype", "lower coverage haplotype")),
        values=(c("darkred", "skyblue")),
    ) +
    scale_x_continuous("Coverage", limits=c(0,100), breaks=seq(0,100,25)) +
    theme_minimal_grid() +
    facet_col(~is_dup, scale="free_y") + 
    theme(legend.position="top")
my_ggsave("Figures/COLO-available-template-total-coverage.pdf") 

```




#
# do hap specific peaks in BL change in Tumor
#

# data
```{r}
vcf="data/PS00388_COLO829BL_1.deepvariant.vcf.gz"
vcf_bl = fread(cmd=glue("bcftools view -f 'PASS' -i 'GT=\"HET\"' {vcf} | grep -v ^##"))  %>% 
    mutate(start = POS-1) %>%
    rename(
        chrom="#CHROM",
        end=POS
    )

colo_hap_peaks = my_read_bed("results/PS00338_COLO829BL_1/hap1-vs-hap2/FIRE.hap.differences.bed") %>%
    filter(p_adjust <= 0.05) %>%
    bed_map(
        vcf_bl,
        is_het = n() > 0,
    ) %>%
    replace_na(list(is_het=FALSE))
    
colo_t_peaks = my_read_bed("results/PS00357_COLO829T_1/FDR-peaks/FDR-FIRE-peaks.bed.gz") 

#colo_tumor_track = my_read_bed("results/PS00357_COLO829T_1/FDR-peaks/FDR.track.bed.gz") %>%
 #   filter(coverage > 0, coverage_H1 > 0, coverage_H2>0) 

colo_hap_peaks %>%
    bed_map(colo_t_peaks,
        t_peak = n() > 0,
    ) %>%
    replace_na(list(t_peak=FALSE)) %>%
    filter(t_peak) 


```

# intersection + analysis
```{r}
#tumor_cov = mean(colo_tumor_track$coverage)
min_hap_cov = 10
filtered_colo_tumor_track = colo_t_peaks %>%
    filter(coverage_H1 > min_hap_cov, coverage_H2 > min_hap_cov) %>%
    mutate(
        hap1_per = fire_coverage_H1/coverage_H1,
        hap2_per = fire_coverage_H2/coverage_H2,
        diff = hap1_per - hap2_per, 
    ) 
sum(is.na(filtered_colo_tumor_track$diff))

colo_compare_df = colo_hap_peaks %>%
    bed_map(
        filtered_colo_tumor_track, 
        tumor_diff = diff[which.max(abs(diff))],
    ) %>%
    bed_map(
        imprinted, 
        is_imprinted = n()>0,
    ) %>%
    replace_na(list(is_imprinted=FALSE)) %>%
    filter(!is.na(tumor_diff)) %>%
    data.table()

asdf = colo_compare_df %>%
    select(chrom, start, end, diff, tumor_diff, is_imprinted, is_het) %>%
    mutate(
        is_het = case_when(
            is_het ~ "Overlaps heterozygous variant",
            TRUE ~ "Homozygous",
        ),
        diff_diffs = abs(diff - tumor_diff),
        large_t_diff = diff_diffs > 0.75,
    ) %>%
    ggplot(aes(x=diff, y=tumor_diff, color=is_imprinted)) +
    geom_abline(slope=1, intercept=0, color="red", linetype="dashed") +
    geom_abline(slope=1, intercept=-0.75, color="gray", linetype="dashed") +
    geom_abline(slope=1, intercept=0.75, color="gray", linetype="dashed") +
    geom_point(alpha=0.75, size=0.5) +
    geom_smooth(method="lm", se=FALSE, alpha=0.1, linewidth=0.1) +
    geom_text_repel(
        data = . %>% filter(large_t_diff),
        aes(
            label=paste0(chrom, ":", start, "-", end),
        ),
        size = 1,
        min.segment.length = 0,
    ) + 
    stat_cor(size=2) + 
    facet_row(~is_het) +
    scale_y_continuous(
        "Percent accessibility difference in tumor peaks\n(filtered for peaks with 10x coverage in both haplotypes)",
        limits=c(-1,1), labels=percent) +
    scale_x_continuous(
        "Percent accessibility difference in blood haplotype-specific peaks",
        limits=c(-1,1), labels=percent) +
    scale_color_manual(
        "Imprinted",
        values=c("black", "darkorange"),
    )+
    # force equal aspect ratio
    coord_fixed() +
    theme_minimal_grid(font_size=6) +
    theme(legend.position="top")
my_ggsave("Figures/COLO-BL-Tumor-hap-peaks.pdf", height=3, width=5)
```
