```{r}
source("Rscripts/utils.R")
source("Rscripts/Fire-peak-df.R")
```
```{r}
vcf="data/PS00388_COLO829BL_1.deepvariant.vcf.gz"
vcf_bl = fread(cmd=glue("bcftools view -f 'PASS' -i 'GT=\"HET\"' {vcf} | grep -v ^##"))  %>% 
    mutate(start = POS-1) %>%
    rename(
        chrom="#CHROM",
        end=POS
    )

bl_peaks = my_read_bed("results/PS00338_COLO829BL_1/hap1-vs-hap2/FIRE.hap.differences.bed")
bl_peaks_2 = my_read_bed("results/PS00356_COLO829BL_2/hap1-vs-hap2/FIRE.hap.differences.bed")
bl_peaks
```

```{r}
bl_peaks_hap = bl_peaks %>% filter( p_adjust <= 0.05 & chrom != "chrX" ) 
bl_peaks_hap_2 = bl_peaks_2 %>% filter( p_adjust <= 0.05 & chrom != "chrX" )
z=bed_merge(
    bind_rows(bl_peaks_hap, bl_peaks_hap_2)
    ) %>%
    bed_map(
        bl_peaks,
        diff = diff[which.max(abs(diff))]
    ) %>%
    bed_map(
        bl_peaks_2,
        diff_2 = diff[which.max(abs(diff))]
    )
sum(is.na(z$diff_2))

z %>%   
    ggplot(aes(diff, diff_2)) +
    geom_hex(bins=50) + scale_fill_distiller(palette="Spectral") +
    geom_abline(slope=1, intercept=0, color="red") +
    stat_cor() +
    xlab("Actuation difference in haplotype-selective peaks (hap1-hap2)\nCOLO829 Blood replicate 1") +
    ylab("Actuation difference in haplotype-selective peaks (hap1-hap2)\nCOLO829 Blood replicate 2") +
    my_grid()+
    theme(
        aspect.ratio=1
    )
my_ggsave("Figures/COLO-hap-selective-replicate-diff-correlation.pdf", width=3, height=3)
```

```{r}
m6a1 = fread(cmd = "zcat /mmfs1/gscratch/stergachislab/mvollger/projects/FIREv2.0/COLO-asm/results/PS00338_DSA/pbmm2/*.tbl.gz  | rg m6a_ratio")
m6a2 = fread(cmd = "zcat /mmfs1/gscratch/stergachislab/mvollger/projects/FIREv2.0/COLO-asm/results/PS00356_DSA/*/*.tbl.gz  | rg m6a_ratio")
z = bind_rows(
    list(
        m6a1=m6a1,
        m6a2=m6a2
    ),
    .id="source"
) 
colnames(z) = c("source", "stat", "value", "count")
z
# geom step histogram of m6a ratios 
z %>%
    ggplot(aes(value, weight=count, color=source)) +
    geom_step(stat="bin", bins=100) +
    scale_x_continuous("m6A ratio") +
    scale_y_continuous("Number of peaks") +
    my_grid() +
    theme(
        aspect.ratio=1
    )
my_ggsave("Figures/COLO-hap-selective-m6a-ratio-histogram.pdf", width=3, height=3)

z %>%
    group_by(
        source
    ) %>%
    summarise(
        totalw = sum(value*count),
        count = sum(count),
        mean = 100*totalw/count
    )
4.49/3.81
3.81/4.49
```



```{r}
set.seed(1) 
nrow(vcf_bl)/1e6

hap_peaks=bl_peaks %>%
    mutate(
        autosome = case_when(
            chrom %in% FAI_NO_SEX$name ~ "Autosome",
            chrom == "chrX" ~ "X",
            TRUE ~ "Other"    
        ),
    ) %>%
    filter(!is.na(diff)) %>%
    filter(p_adjust<=0.05 & autosome=="Autosome" ) %>%
    mutate(
        imprinted=FALSE
    ) %>%
    data.table

rand_hap_peaks = bed_shuffle(hap_peaks, FAI)


hap_peaks_with_null = bind_rows(
    list(real=hap_peaks, shuffle=rand_hap_peaks),
    .id="type"
    ) %>%
    bed_map(vcf_bl,
        n_SNVs = n(),
    ) %>% 
    replace_na(list(
        n_SNVs=0
    )) %>%
    mutate(
        str_imprinted = case_when(
            imprinted == TRUE ~ "imprinted",
            TRUE ~ "not-imprinted",
        ),
        n_variants = n_SNVs,
        has_variants = case_when(
            n_variants > 0 ~ "With variants",
            TRUE ~ "Without variants",
        ),
    )
hap_peaks_with_null %>% group_by(type) %>% summarise(sum(imprinted), n())

p_sum = hap_peaks_with_null %>% group_by(type) %>%
    summarise(
        n_with_variants = sum(n_variants > 0),
        n_without_variants = sum(n_variants == 0)
    ) %>%
    ungroup %>%
    pivot_wider(names_from=type, values_from=starts_with("n_")) %>%
    rowwise %>%
    mutate(
        p_value = fisher.test(matrix(
            c(n_with_variants_real, n_without_variants_real, n_with_variants_shuffle, n_without_variants_shuffle),
            ncol=2),
            alternative="greater"
            )$p.value,        
    )
p_sum
```

```{r}
hap_peaks_with_null %>%
    filter(type=="real") %>%
    ggplot(
        aes(
            x=pmin(n_variants, 10),
        )
    ) +
    geom_histogram(
        aes(),#fill=str_imprinted),
        binwidth=1,
        #color="black",
        width=0.4,
    ) +
    stat_bin(
        aes(
            y = after_stat(count),
            #fill=str_imprinted,
            label=ifelse(..count.. > 0, ..count.., ""),
        ),
        #color="black",
        vjust=1,
        binwidth=1,
        geom="text_repel",
        min.segment.length=unit(1000, "lines"),
        direction="y",
        size=1.5,
    ) +
    scale_x_continuous(
        glue("# of heterozygous variants (SNVs, indels)\nper hap-selective peak (n={comma(nrow(hap_peaks))} autosomal peaks)"),
        breaks=seq(0,10),
        labels=c(seq(0,9), "10+")
    ) +
    scale_y_continuous("Number of peaks") +
    my_hgrid() +
    #scale_fill_manual("",
    #    values=c("imprinted"="darkorange","not-imprinted"="darkgrey")
    #) +
    ggtitle(glue(
    "Fisher's exact one-sided p-value: {scientific(p_sum$p_value)}\n",
    "Real data: with {p_sum$n_with_variants_real} without {p_sum$n_without_variants_real};   ",
    "Shuffled data: with {p_sum$n_with_variants_shuffle} without {p_sum$n_without_variants_shuffle}"
    )) +
    theme(
        legend.position="top",
        plot.title = element_text(size = 5, face = "bold"),
    ) 

my_ggsave("Figures/COLO-hap-selective-n-variants-per-peak.pdf", height=3, width=3)
```