---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
# get a fire peak data table called `df`
source("Rscripts/utils.R")
source("Rscripts//Fire-peak-df.R")
getmode <- function(v) {
   uniqv <- unique(v)
   uniqv[which.max(tabulate(match(v, uniqv)))]
}
library(rlang)
#library(gmm)
# Load the package
library(mclust)
```

```{r}
### UNUSED ###

SM="COLO_T_2_PS00_418_451_488"
all_covs = my_read_bed(glue("results/{SM}/FDR-peaks/FDR.track.bed.gz")) %>%
    # require 95% phasing
    filter(
        `chrom` != "chrX",
        `chrom` != "chrY",
    ) %>%
    filter(coverage * 0.90 <= coverage_H1 + coverage_H2) %>%
    filter(
        pmin(coverage_H1, coverage_H2) > 1,
    ) %>%
    sample_n(1e6) %>%
    data.table

model_cov = all_covs %>% filter(
    coverage < 200,
    coverage > 30
)

pdf("tmp.pdf")
hist(model_cov$coverage_H1, breaks=100, xlim=c(0,100))
hist(model_cov$coverage_H2, breaks=100, col="red", xlim=c(0,100))
hist(model_cov$coverage, breaks=100, col="blue", xlim=c(0,100))
hist(model_cov$coverage_H1 + model_cov$coverage_H2, breaks=100, col="blue", xlim=c(0,100))
dev.off()
nrow(fire)

m = Mclust(c(model_cov$coverage_H1, model_cov$coverage_H2), G=seq(2))
m$parameters$mean
```

```{r}
# Tumor PS00357
fire = my_read_bed(glue("results/{SM}/FDR-peaks/FDR-FIRE-peaks.bed.gz")) %>%
    # require 95% phasing
    filter(
        `chrom` != "chrX",
        `chrom` != "chrY",
    ) %>%
    filter(coverage * 0.9 <= coverage_H1 + coverage_H2) %>%
    filter(
        pmin(coverage_H1, coverage_H2) > 1,
    ) %>%
    mutate(
        high_cov_hap = case_when(
            coverage_H1 >= coverage_H2 ~ coverage_H1,
            TRUE ~ coverage_H2
        ),
        high_cov_hap_fire = case_when(
            fire_coverage_H1 >= fire_coverage_H2 ~ fire_coverage_H1,
            TRUE ~ fire_coverage_H2
        ),
        low_cov_hap = case_when(
            coverage_H1 < coverage_H2 ~ coverage_H1,
            TRUE ~ coverage_H2
        ),
        low_cov_hap_fire = case_when(
            fire_coverage_H1 < fire_coverage_H2 ~ fire_coverage_H1,
            TRUE ~ fire_coverage_H2
        ),
    )
sd_range = 3
mean_hap_cov = mean(fire$coverage_H1 + fire$coverage_H2)/2
lower_hap_cov_limit = mean_hap_cov - sd_range * sqrt(mean_hap_cov)
upper_hap_cov_limit = mean_hap_cov + sd_range * sqrt(mean_hap_cov)
print(glue("Mean coverage: {mean_hap_cov}\tLower limit: {lower_hap_cov_limit}\tUpper limit: {upper_hap_cov_limit}"))

norm_vs_dup_df_wide = fire %>% 
    mutate(
        normal_low_cov = (low_cov_hap > lower_hap_cov_limit), #& (low_cov_hap < upper_hap_cov_limit),
        is_dup = (high_cov_hap >= 2*low_cov_hap),
    ) %>%
    mutate(
        percent_accessible_dup = 100*high_cov_hap_fire / high_cov_hap,
        percent_accessible_norm = 100*low_cov_hap_fire / low_cov_hap,
    ) %>%
    mutate(
        norm_minus_dup = percent_accessible_norm - percent_accessible_dup 
    )
    
mean(norm_vs_dup_df_wide$normal_low_cov)
mean(norm_vs_dup_df_wide$is_dup)    
    
norm_vs_dup_df = norm_vs_dup_df_wide %>%
    pivot_longer(
        cols = c("percent_accessible_dup", "percent_accessible_norm"),
        names_to = "type",
        values_to = "percent"
    )

norm_vs_dup_df_coverage = norm_vs_dup_df_wide %>%
    pivot_longer(
        cols = c("high_cov_hap", "low_cov_hap"),
        names_to = "type",
        values_to = "coverage_hap"
    )

table(norm_vs_dup_df_wide[,c("is_dup", "normal_low_cov")])
```

```{r}
```

```{r}
norm_vs_dup_df_wide %>%
    rename(
        `lower coverage haplotype` = low_cov_hap,
        `higher coverage haplotype` = high_cov_hap,
        `haplotype 1` = coverage_H1,
        `haplotype 2` = coverage_H2,
    ) %>%
    pivot_longer(
        cols = c("lower coverage haplotype", "higher coverage haplotype", "haplotype 1", "haplotype 2"),
        names_to = "type",
        values_to = "coverages"
    ) %>%
    select(type, coverages, is_dup, normal_low_cov) %>%
    mutate(
        per = rank(coverages)/n(),
    ) %>%
    filter(per < 0.99) %>%
    ggplot(
        aes(
            x=coverages,
            fill=is_dup, #& normal_low_cov,
        )
    ) +
    scale_fill_manual(
        "Two fold coverage difference\nbetween H1 and H2",
        values=c("skyblue", "darkred"),
    ) +
    geom_histogram(binwidth=1, position="stack", alpha=0.7) +
    scale_x_continuous("Coverage") +
    facet_col(~type) +
    my_grid() +
    theme(legend.position="top")
my_ggsave("Figures/COLO-available-template-total-coverages.pdf", height=3, width=3) 
```

```{r}
```

```{r}
norm_vs_dup_df %>%
    filter(is_dup, normal_low_cov) %>%
    ggplot(aes(fill=type, x=percent)) +
    #geom_histogram(binwidth=2, position="identity", alpha=0.7) +
    geom_histogram(
        aes(y=100*after_stat(density)),
        binwidth=5,
        alpha=0.7,
        position="dodge",
    ) +
    scale_fill_manual("",
        labels=rev(c("Duplicated haplotype", "Normal haplotype")),
        values=rev(c("darkred", "skyblue")),
    ) +
    scale_x_continuous("Percent accessible", limits=c(0, 100)) +
    my_grid() +
    theme(legend.position="top")
my_ggsave("Figures/COLO-available-template.pdf", height=3, width=3) 

norm_vs_dup_df_wide %>%
    filter(normal_low_cov, is_dup) %>%
    mutate(
        nrank = rank(1/abs(norm_minus_dup))
    ) %>%
    ggplot(
        aes(
            x=percent_accessible_norm,
            y=percent_accessible_dup,
        )
    ) + 
    geom_hex(bins=30) + scale_fill_distiller("", palette = "Spectral", trans="log10") +
    geom_text_repel(
        data = . %>% filter(nrank <= 6),
        aes(
            label=paste0(chrom, ":", comma(start), "-", comma(end)),
        ),
        color="black",
        min.segment.length=0,
        segment.size=0.25,
        arrow = arrow(length = unit(0.05, 'cm'), type = 'closed'),
        size = 1
    ) + 
    #geom_smooth(method="lm", se=FALSE, color="black", size=0.5) +
    geom_abline(slope=1, intercept=0, color="black", linetype="dashed") +
    stat_cor()+
    scale_x_continuous("Percent accessible (low coverage haplotype)", limits=c(0,100)) +
    scale_y_continuous("Percent accessible (high coverage haplotype)", limits=c(0,100)) +
    coord_fixed() +
    #facet_col(~is_dup) + 
    my_grid()
my_ggsave("Figures/COLO-available-template-scatter.pdf", height=3, width=3) 


norm_vs_dup_df_wide %>%
    filter(normal_low_cov) %>%
    ggplot(aes(x=norm_minus_dup, fill=is_dup)) + 
    geom_density(alpha=0.4,
        #aes(y = after_stat(scaled))
    ) +
    scale_x_continuous("Percent accessibility difference\n(low coverage haplotype - high coverage haplotype)", limits=c(-100,100)) +
    theme_minimal_grid() +
    #facet_col(~is_dup)+
    scale_fill_manual("",
        labels=c("Regions with normal coverage", "Regions with one duplicated haplotype"),
        values=c("skyblue", "darkred"),
    ) +
    theme(legend.position="top")
my_ggsave("Figures/COLO-available-template-diff.pdf")


# plot the coverage of the duplicated haplotype vs the normal haplotype
norm_vs_dup_df_coverage %>%
    mutate( 
        is_dup = case_when(
            is_dup ~ "One haplotype has >2 fold more coverage",
            TRUE ~ "Both haplotypes have similar coverage",
        )
    ) %>%
    #filter(is_dup) %>%
    ggplot(aes(fill=type, x=coverage_hap)) +
    geom_density(
        #aes(y = after_stat(scaled)),
        alpha=0.75, adjust=.6
    ) +
    scale_fill_manual("",
        labels=(c("higher coverage haplotype", "lower coverage haplotype")),
        values=(c("darkred", "skyblue")),
    ) +
    scale_x_continuous("Coverage", limits=c(0,150), breaks=seq(0,250,25)) +
    theme_minimal_grid() +
    facet_col(~is_dup, scale="free_y") + 
    theme(legend.position="top")
my_ggsave("Figures/COLO-available-template-total-coverage.pdf") 

```




#
# do hap specific peaks in BL change in Tumor
#

# data
```{r}
vcf="data/PS00388_COLO829BL_1.deepvariant.vcf.gz"
vcf_bl = fread(cmd=glue("bcftools view -f 'PASS' -i 'GT=\"HET\"' {vcf} | grep -v ^##"))  %>% 
    mutate(start = POS-1) %>%
    rename(
        chrom="#CHROM",
        end=POS
    )

colo_hap_peaks = my_read_bed("results/PS00338_COLO829BL_1/hap1-vs-hap2/FIRE.hap.differences.bed") %>%
    filter(p_adjust <= 0.05)
    
colo_t_peaks = my_read_bed(glue("results/{SM}/FDR-peaks/FDR-FIRE-peaks.bed.gz")) %>%
    mutate(
        diff = fire_coverage_H1/coverage_H1 - fire_coverage_H2/coverage_H2,
    )

colo_t_passage_a = my_read_bed("results/PS00357_COLO829T_1/FDR-peaks/FDR-FIRE-peaks.bed.gz") %>%
    mutate(
        diff = fire_coverage_H1/coverage_H1 - fire_coverage_H2/coverage_H2,
    )

rep_2_peaks = my_read_bed(glue("results/PS00356_COLO829BL_2/FDR-peaks/FDR-FIRE-peaks.bed.gz"))
#colo_tumor_track = my_read_bed("results/PS00357_COLO829T_1/FDR-peaks/FDR.track.bed.gz") %>%
 #   filter(coverage > 0, coverage_H1 > 0, coverage_H2>0) 

other_colo_peaks = bind_rows(
    list(
        "Tumor passage A"=colo_t_passage_a,
        "Tumor passage B"=colo_t_peaks,
        "Blood replicate 2"=rep_2_peaks
    ),
    .id = "source"
)
```

# intersection + analysis
```{r}
min_hap_cov = 10
filtered_colo_tumor_track = other_colo_peaks %>%
    filter(coverage_H1 > min_hap_cov, coverage_H2 > min_hap_cov) %>%
    mutate(
        hap1_per = fire_coverage_H1/coverage_H1,
        hap2_per = fire_coverage_H2/coverage_H2,
        diff = hap1_per - hap2_per, 
    ) 
sum(is.na(filtered_colo_tumor_track$diff))

colo_compare_df = filtered_colo_tumor_track %>%
    bed_map(
        colo_hap_peaks,
        blood_diff = diff[which.min(abs(diff))],
        nn = n(),
    ) %>%
    filter(!is.na(blood_diff)) %>%
    bed_map(
        imprinted, 
        is_imprinted = n()>0,
    ) %>%
    bed_map(
        vcf_bl,
        is_het = n() > 0,
    ) %>%
    replace_na(list(is_het=FALSE, is_imprinted=FALSE))
sum(colo_compare_df$is_het)
table(colo_compare_df$nn)

asdf = colo_compare_df %>%
    filter(source != "Tumor passage A") %>%
    select(chrom, start, end, diff, blood_diff, is_imprinted, is_het, source) %>%
    #expand_grid(fake=seq(1)) %>%
    filter(!(is_het & is_imprinted)) %>%
    mutate(
        is_het = case_when(
            is_het ~ "Overlaps heterozygous variant",
            TRUE ~ "Homozygous",
        ),
        diff_diffs = abs(diff - blood_diff),
        large_t_diff = diff_diffs > 0.5,
    ) %>%
    ggplot(aes(x=blood_diff, y=diff, color=is_imprinted)) +
    geom_abline(slope=1, intercept=0, color="red", linetype="dashed") +
    geom_abline(slope=1, intercept=-0.5, color="gray", linetype="dashed") +
    geom_abline(slope=1, intercept=0.5, color="gray", linetype="dashed") +
    geom_point(alpha=0.75, size=0.5) +
    geom_smooth(method="lm", se=FALSE, alpha=0.1, linewidth=0.1) +
    geom_text_repel(
        data = . %>% filter(large_t_diff),
        aes(
            label=paste0(chrom, ":", start, "-", end),
        ),
        size = 1,
        min.segment.length = 0,
    ) + 
    stat_cor(size=2) + 
    facet_grid(source~is_het) +
    scale_y_continuous(
        "Percent accessibility difference in tumor/replicate peaks\n(filtered for peaks with 10x coverage in both haplotypes)",
        limits=c(-1,1), labels=percent) +
    scale_x_continuous(
        "Percent accessibility difference in blood haplotype-specific peaks",
        limits=c(-1,1), labels=percent) +
    scale_color_manual(
        "Imprinted",
        values=c("black", "darkorange"),
    )+
    # force equal aspect ratio
    coord_fixed() +
    theme_minimal_grid(font_size=6) +
    theme(legend.position="top")
my_ggsave("Figures/COLO-BL-Tumor-hap-peaks.pdf", height=4, width=4)
```

same plot as above but passage A vs B for tumor
```{r}
t_a = colo_compare_df %>% 
    filter(source=="Tumor passage A") %>%
    select(chrom, start, end, diff, is_imprinted, is_het, source) %>%
    bed_map(
        colo_compare_df %>% filter(source=="Tumor passage B"),
        diff_b = diff[which.min(abs(diff))],
        nn = n(),
    ) 

asdf = t_a %>%
    #expand_grid(fake=seq(1)) %>%
    filter(!(is_het & is_imprinted)) %>%
    mutate(
        is_het = case_when(
            is_het ~ "Overlaps heterozygous variant",
            TRUE ~ "Homozygous",
        ),
        diff_diffs = abs(diff - diff_b),
        large_t_diff = diff_diffs > 0.5,
    ) %>%
    ggplot(aes(x=diff_b, y=diff, color=is_imprinted)) +
    geom_abline(slope=1, intercept=0, color="red", linetype="dashed") +
    geom_abline(slope=1, intercept=-0.5, color="gray", linetype="dashed") +
    geom_abline(slope=1, intercept=0.5, color="gray", linetype="dashed") +
    geom_point(alpha=0.75, size=0.5) +
    geom_smooth(method="lm", se=FALSE, alpha=0.1, linewidth=0.1) +
    geom_text_repel(
        data = . %>% filter(large_t_diff),
        aes(
            label=paste0(chrom, ":", start, "-", end),
        ),
        size = 1,
        min.segment.length = 0,
    ) + 
    stat_cor(size=2) + 
    facet_grid(source~is_het) +
    scale_y_continuous(
        "Percent accessibility difference in tumor passage A",
        limits=c(-1,1), labels=percent) +
    scale_x_continuous(
        "Percent accessibility difference in tumor passage B",
        limits=c(-1,1), labels=percent) +
    scale_color_manual(
        "Imprinted",
        values=c("black", "darkorange"),
    )+
    # force equal aspect ratio
    coord_fixed() +
    theme_minimal_grid(font_size=6) +
    theme(legend.position="top")

my_ggsave("Figures/COLO-TumorA-TumorB-hap-peaks.pdf", height=3, width=5)
```


Thought re COLO829T. Are the 0 variant BL-Hap-specific sites enriched for sites that show more haplotype-specific heterogeneity (i.e. less well buffered).
Make a histogram of the absolute(haplotype difference in COLO829T for all peaks that are present in BOTH COLO829BL and COLO829T)
Add to histogram absolute(haplotype difference in COLO829T for peaks that are COLO829BL haplotype-specific and also present in COLO829T and lack a variant)
(edited)
you have already showed that there is no correlation between the COLO829T and COLO829BL haplotype-specificity, but this would show that these sites are "less buffered" such that they are more likely to diverge over time in their haplotype-specificity compared to other peaks around the genome

```{r}
# get the peaks that are present in both COLO829BL and COLO829T
colo_bl = my_read_bed("results/PS00338_COLO829BL_1/hap1-vs-hap2/FIRE.hap.differences.bed") %>%
    mutate(
        frac_acc = fire_coverage/coverage,
    )


min(colo_bl %>% pull(frac_acc))
min_frac_acc = min(colo_bl %>% filter(p_adjust <= 0.05) %>% pull(frac_acc))

# get the peaks that are present in both COLO829BL and COLO829T
col_bl_v_t_df = colo_bl %>%
    bed_map(
        colo_t_peaks %>%
            filter(coverage_H1 >= 10, coverage_H2 >= 10),
        is_tumor = n() > 0,
        tumor_diff = diff[which.min(abs(diff))],
        n_tumor = n(),
    ) %>%
    filter(!is.na(is_tumor)) %>%
      bed_map(
        imprinted, 
        is_imprinted = n()>0,
    ) %>%
    bed_map(
        vcf_bl,
        is_het = n() > 0,
    ) %>%
    replace_na(list(is_het=FALSE, is_imprinted=FALSE))

ppd = col_bl_v_t_df %>%
    filter(!is_het) %>%
    mutate(
        type_diff = abs(diff - tumor_diff),
        hap_selective = case_when(
            p_adjust <= 0.05 ~ "Haplotype selective",
            TRUE ~ "Non-haplotype selective",
        )
    ) 
table(ppd$hap_selective)
colo_compare_df %>% filter(!is_het, source=="Tumor") %>% nrow()
ppd %>%
    ggplot(aes(x=type_diff, fill=hap_selective)) +
    geom_histogram(
        aes(y=after_stat(density)),
        binwidth=0.05, alpha=0.75,
        color="black",
        position="identity",
    ) +
    #geom_density(alpha=0.75)+
    scale_fill_manual(
        "",
        values=c("darkred","skyblue"),
    ) +
    geom_text(
        data = . %>% dplyr::group_by(hap_selective) %>% dplyr::summarise(stat=mean(type_diff)),
        aes(x=stat, y=-.2, 
            label=glue("mean={100*round(stat, 3)}"), color=hap_selective),
        size=2.5,
    ) +
    scale_x_continuous(
        glue(
            "abs[(Blood H1 - H2 accessibility) - (Tumor H1 - H2 accessibility)]\n",
            "(shared homozygous peaks with at least 10x coverage in both haplotypes)"
        ),
        label=percent,
        limits=c(0,1),
    ) +
    scale_color_manual(
        "",
        values=c("darkred", "skyblue"),
    ) +
    #facet_col(~hap_selective, scale="free_y") +
    my_grid() +
    theme(
        legend.position="top"
    )
my_ggsave("Figures/COLO-BL-Tumor-histogram-of-hap-difference.pdf", height=3, width=4)
```