---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
# get data
source("Rscripts/Fire-peak-df.R")
```

```{r}
f1 = "results/GM12878/hap1-vs-hap2/FIRE.hap.differences.bed"
f2 = "results/PS00338_COLO829BL_1/hap1-vs-hap2/FIRE.hap.differences.bed"
f3 = "results/PS00356_COLO829BL_2/hap1-vs-hap2/FIRE.hap.differences.bed"
f4 = "results/PS00381/hap1-vs-hap2/FIRE.hap.differences.bed"
f5 = "results/PS00382/hap1-vs-hap2/FIRE.hap.differences.bed"
f6 = "results/PS00383/hap1-vs-hap2/FIRE.hap.differences.bed"
f7 = "results/PS00384/hap1-vs-hap2/FIRE.hap.differences.bed"
f8 = "results/PS00272/hap1-vs-hap2/FIRE.hap.differences.bed"
f9 = "results/PS00321/hap1-vs-hap2/FIRE.hap.differences.bed"
f10 = "results/PS00327/hap1-vs-hap2/FIRE.hap.differences.bed"
f11 = "results/UDN318336_PS00189/hap1-vs-hap2/FIRE.hap.differences.bed"
f12 = "results/UDN633333_PS00115/hap1-vs-hap2/FIRE.hap.differences.bed"
df1=my_read_bed(f1)
df2=my_read_bed(f2)
df3=my_read_bed(f3)
df4=my_read_bed(f4)
df5=my_read_bed(f5)
df6=my_read_bed(f6)
df7=my_read_bed(f7)
df8=my_read_bed(f8)
df9=my_read_bed(f9)
df10=my_read_bed(f10)
df11=my_read_bed(f11)
df12=my_read_bed(f12)

df_all = bind_rows(
    list(
        GM12878=df1,
        PS00338=df2,
        PS00356=df3,
        PS00381=df4,
        PS00382=df5,
        PS00383=df6,
        PS00384=df7,
        PS00272=df8,
        PS00321=df9,
        PS00327=df10,
        UDN318336_PS00189=df11,
        UDN633333_PS00115=df12
        ),
    .id="sample"
) %>% 
    bed_map(SDs,
        is_SD = n() > 0
    ) %>%
    bed_map(ALTs,
        is_Alt = n() > 0
    ) %>%
    bed_map(imprinted,
        is_imprinted = n() > 0
    ) %>%
    replace_na(
        list(
            is_SD=F,
            is_Alt=F,
            is_imprinted=F
        )
    ) %>%
    data.table

samples = unique(df_all$sample)
colors = c(
    "LCL"="darkgreen",
    "CD8+"="darkred",
    "Thyroid"="darkblue",
    "Fibroblast"="darkorange"
    )
sample_type = c(
    "GM12878"="LCL",
    "PS00338"="LCL",
    "PS00356"="LCL",
    "PS00381"="CD8+",
    "PS00382"="CD8+",
    "PS00383"="CD8+",
    "PS00384"="CD8+",
    "PS00272"="Thyroid",
    "PS00321"="Thyroid",
    "PS00327"="Thyroid",
    "PS00388"="Thyroid",
    "UDN318336_PS00189"="Fibroblast",
    "UDN633333_PS00115"="Fibroblast"
)
df_all$sample_type = factor(sample_type[df_all$sample], levels=rev(names(colors)))
df_all$color = colors[df_all$sample_type]
df_all = df_all %>% arrange(sample_type, sample, chrom, start, end)

df = df_all %>%
 filter(p_value <= 0.05) %>%
 filter(abs(diff) >= 0.25) 
```


```{r}
system("rm tmp.pdf")
s = 4
pdf("tmp.pdf", height=2*s, width=8*s)

mhc = "chr6:28,510,120-33,480,577"
mhc.region = toGRanges(data.frame("chr6", 28510120, 33480577))
notch="chr1:146,148,865-146,229,021"
notch.region = toGRanges(data.frame("chr1", 146148865-1e6, 146229021+1e6))

mhc.region

chrs = paste0("chr", seq(1,22))
for(idx in seq(4)){
    ymax=20
    if(idx == 1){
        c = chrs[1:22]
        ws=10e5
        grouping_var = "sample"
    }else if (idx == 2){
        # skip this block
        c = chrs[1:22]
        ws=10e5
        grouping_var = "sample_type"
    } else if (idx == 3){
        ymax=NULL
        c = "chr6"
        zoom.region = mhc.region
        ws = 1e4
    } else if(idx == 4){
        ymax=NULL
        c = "chr1"
        zoom.region = notch.region
        ws = 1e4
    }

    tdf=df[p_value <= 0.05 & `chrom` %in% c] %>%
        arrange(sample_type, sample, .by_group=T)

    gr = GRanges(seqnames=tdf$`chrom`, IRanges(start=tdf$start, end=tdf$end))
    imprinted_gr = GRanges(imprinted[chrom %in% c])

    if(length(c) == 1){
        kp = karyoploteR::plotKaryotype(genome="hg38", plot.type=4, zoom=zoom.region)
    }else{
        kp = karyoploteR::plotKaryotype(genome="hg38", plot.type=4, chromosomes = c)
    }
    
    kpAddBaseNumbers(kp)

    #kpPlotDensity(kp, gr, col="blue", ymax = ymax, r0=0.0, r1=0.5, border=NA, window.size=ws,)
    kpPlotRegions(kp, imprinted_gr, col="darkorange", data.panel=1, r0=-0.05, r1=0.0)

    r0 = 0 # 0.5
    r_width = 1/length(unique(tdf[[grouping_var]])) # * 0.5
    for(var in unique(tdf[[grouping_var]])){
        r1 = r0 + r_width*0.85
        cond = tdf[[grouping_var]] == var
        ttdf = tdf[cond,]
        st = ttdf[[grouping_var]][[1]]
        color = ttdf$color[[1]]
        sample.regions = GRanges(seqnames=ttdf$`chrom`, IRanges(start=ttdf$start, end=ttdf$end))
        
        tkp = kpPlotDensity(
            kp, data=sample.regions, 
            r0=r0, r1=r1, 
            border=transparent(color, 0.5),
            col=color,
            #border=NA,
            data.panel=1,
            #Zymax=ymax,
            window.size=ws,
        )
        print(tkp$latest.plot$computed.values$max.density)
        kpAxis(kp, 
            ymax=tkp$latest.plot$computed.values$max.density,
            r0=r0, r1=r1,
            cex=0.5,
            side="right",
            numticks=2,
        )
        kpAddLabels(kp, labels = st, r0=r0, r1=r1, cex=1)

        r0 = r0 + r_width
        print(nrow(ttdf))
    }
}

dev.off()
system("cp tmp.pdf Figures/differential-peaks-ideogram.pdf")
dev.off()
```











# ~Mbp similarity comparison, intra vs inter hap similarity

```{r}
library(lsa)
AUO = F
if(AUO){
    peak_files = c(
        "results/PS00381/FDR-peaks/FDR-FIRE-peaks.bed.gz",
        "results/PS00382/FDR-peaks/FDR-FIRE-peaks.bed.gz",
        "results/PS00383/FDR-peaks/FDR-FIRE-peaks.bed.gz",
        "results/PS00384/FDR-peaks/FDR-FIRE-peaks.bed.gz"
    )
} else {
    peak_files = c(
        "results/GM12878/FDR-peaks/FDR-FIRE-peaks.bed.gz",
        "results/PS00338_COLO829BL_1/FDR-peaks/FDR-FIRE-peaks.bed.gz"
    )
}
row_files = paste(peak_files, sep = " ", collapse = " ")
all_peaks = fread(cmd=glue("less {row_files} | bedtools sort | bedtools merge"))
colnames(all_peaks) = c("chrom", "start", "end")
all_peaks
```


```{r}
if(AUO){
    PS00381 = fread("results/PS00381/FDR-peaks/FDR.track.coverage.filtered.bed.gz")
    PS00382 = fread("results/PS00382/FDR-peaks/FDR.track.coverage.filtered.bed.gz")
    PS00383 = fread("results/PS00383/FDR-peaks/FDR.track.coverage.filtered.bed.gz")
    PS00384 = fread("results/PS00384/FDR-peaks/FDR.track.coverage.filtered.bed.gz")
    df_list = list(
        PS00381=PS00381,
        PS00382=PS00382,
        PS00383=PS00383,
        PS00384=PS00384
    )
    MIN_COV=5
}else{
    GM12878 = fread("results/GM12878/FDR-peaks/FDR.track.coverage.filtered.bed.gz")
    PS00338 = fread("results/PS00338_COLO829BL_1/FDR-peaks/FDR.track.coverage.filtered.bed.gz")
    df_list = list(
        GM12878=GM12878,
        PS00338=PS00338
    )
    MIN_COV=30
}

AUO_samples = names(df_list)
AUO_samples


AUO_df = bind_rows(
    df_list,
    .id="sample"
    ) %>%
    rename(
        `chrom` = `#chrom`
    ) %>% 
    filter(coverage_H1 > MIN_COV & coverage_H2 > MIN_COV) %>%
    group_by(sample) 
    
has_all_pairs = all_peaks %>%
    expand_grid(sample = AUO_samples) %>%
    group_by(sample) %>%
    bed_map(AUO_df,
        fire_score_H1=max(score_H1),
        fire_score_H2=max(score_H2),
    ) %>%
    pivot_wider(
        names_from = sample, 
        values_from = c(fire_score_H1, fire_score_H2)
    ) %>%
    mutate(total_score = rowSums(across(starts_with("fire_score_")), na.rm=T)) %>%
    filter(!is.na(total_score), total_score>0) %>%
    select(-total_score) %>%
    drop_na()
has_all_pairs
```
```{r}
id_cols = names(head(has_all_pairs) %>% select(!starts_with("fire_score")))
score_cols = names(head(has_all_pairs) %>% select(starts_with("fire_score")))
name_pairs = t(combn(score_cols,2))
#name_pairs = unique(rbind(name_pairs,name_pairs[,ncol(name_pairs):1]))
name_pairs
fire_pairs = list()
for(idx in seq(nrow(name_pairs))){
    name_1 = name_pairs[idx,1]
    name_2 = name_pairs[idx,2]
    cols = c(id_cols, name_1, name_2)
    tdf = has_all_pairs %>%
        select(all_of(cols)) %>%
        filter(!if_any(everything(), is.na)) %>%
        mutate(
            s1 = name_1,
            s2 = name_2
        ) %>%
        rename(
            score_1 = name_1,
            score_2 = name_2
        ) %>%
        separate(col = s1,into = c("a", "b", "hap_1", "sample_1")) %>%
        separate(col = s2,into = c("b", "a", "hap_2", "sample_2")) %>%
        select(-a,-b)

    fire_pairs[[idx]] = tdf
}
fire_pairs = bind_rows(fire_pairs) %>%
    arrange(chrom, start) %>%
    mutate(
        same_sample = sample_1 == sample_2,
        group = floor(start/1e6)
    )
fire_pairs
```

```{r}
library(runner)
library(parallel)
ws = 5e5

roll_similarity = function(df, min_tests = 10){
    a = min(df$start)
    b = max(df$end)
    n = nrow(df)
    if(n < min_tests){
        print(glue("{a} - {b}, n = {n}"))
        return(NA)
    }
    cosine(df$score_1, df$score_2)[1]
}

cl <- makeForkCluster(16)
n_windows = 100
similarity_df = fire_pairs %>%
    group_by(sample_1, hap_1, sample_2, hap_2) %>%
    arrange(chrom, start, end) %>%
    mutate(
        similarity = runner(
            .,
            #k=ws, idx="start", lag=ws/2,
            k = n_windows, lag = n_windows/2,
            f = roll_similarity,
            na_pad = NA,
            cl=cl,
        ),
        group_start = lag(start, n_windows/2),
        group_end = lead(end, n_windows/2),
    ) %>% mutate(
        color = case_when(
            sample_1 == sample_2 ~ sample_1,
            TRUE ~ "different"
        )
    ) %>%
    drop_na() %>% 
    arrange(sample_1, chrom, start, end) %>%
    dplyr::group_by(sample_1, chrom, start, end) %>%
    dplyr::arrange(-similarity) %>%
    mutate(
        rank = row_number(),
    ) %>%
    arrange(chrom,start,end,sample_1, -similarity) %>% 
    ungroup()
stopCluster(cl)
similarity_df

unique(similarity_df[c("chrom", "start", "end")])
```

```{r}
similarity_df %>%
    ggplot(
        aes(
            color=color,
            alpha=same_sample,
            group=paste0(sample_1,sample_2,hap_1,hap_2)),
            size = same_sample,
    ) +
    geom_line(aes(y=similarity, x=start)) + 
    geom_rug() +
    scale_x_continuous("Genomic position", label=comma) + 
    scale_alpha_manual("", values=c(0.35, 0.8)) +
    scale_color_manual("", values=c("black", "red", "blue")) +
    #facet_zoom(xlim = c(28510120, 33480577), zoom.size = 1) +
    facet_col(~chrom)+ 
    theme_cowplot() +
    theme(legend.position="top")
#my_ggsave("Figures/MHC-different-epi-haplotypes.pdf", width=22, height=8)
my_ggsave("Figures/All-different-epi-haplotypes.pdf", width=22, height=22)
```


```{r}
get_rank = function(z){
    s = unique(z$sample_1)
    is_s1 = z$sample_1 == s[1] | z$sample_2 == s[1]
    is_s2 = z$sample_1 == s[2] | z$sample_2 == s[2]

    rank1 = seq(nrow(z[is_s1,]))
    rank2 = seq(nrow(z[is_s2,]))
    z$rank[is_s1] = rank1
    z$rank[is_s2] = rank2
    z[z$same_sample,]
}

z=similarity_df %>% 
    ungroup() %>%
    arrange(chrom, start, end) %>%
    dplyr::group_by(chrom, start, end) %>%
    dplyr::arrange(-similarity) %>%
    group_modify( ~ get_rank(.x)) %>%
    #filter(same_sample, rank > 0)
    as.tibble() 


windows_100k = bed_makewindows(FAI[chrom %in% unique(similarity_df$chrom)], 1e5) %>% expand_grid(sample_1 = AUO_samples) %>% group_by(sample_1)
windows_100k

rank_100k = bed_map(windows_100k, z %>% group_by(sample_1),
        max_rank=max(rank),
        min_rank=min(rank),
        mean_rank=mean(rank),
        group_start=mean(group_start),
        group_end=mean(group_end),

    ) %>% bed_map(imprinted,
        imprinted= n() > 0, 
    ) %>%
    bed_map(ALTs,
        is_Alt = n() > 0
    ) %>%
    mutate(chrom = factor(chrom, levels=unique(FAI$chrom))) %>%
    replace_na(
        list(
            imprinted=F,
            is_Alt=F
        )
    ) 

# find the regions where both samples are at least rank 2
both_rank_2 = rank_100k %>% 
    drop_na() %>%
    group_by(chrom, start, end, group_start, group_end) %>%
    dplyr::summarize(
        both_rank_2 = sum(mean_rank >= 2) == 2,
        imprinted = sum(imprinted) > 0,
        is_Alt = sum(is_Alt) > 0,
        n = n()
    ) %>%
    bed_map(SDs,
        is_SD = n() > 0
    ) %>%
    replace_na(
        list(is_SD=F)
    ) %>%
    filter(both_rank_2) 
table(both_rank_2$is_Alt)
table(both_rank_2$is_SD)
both_rank_2 %>% 
    dplyr::summarize(
        SD = sum(is_SD),
        NOT_SD = sum(!is_SD),
        total = n(),
        percent_SD = SD/total * 100, 
    )

rank_100k %>%
    filter(mean_rank >= 2) %>%
    dplyr::group_by(sample_1) %>% 
    dplyr::summarize(
        x = sum(end-start)/1e6
    )
seen_mbp = sum(both_rank_2$end - both_rank_2$start)/1e6
seen_mbp
expected_mbp = 32e6/3e9 * 40e6/3e9 * 3e9 * 1/1e6
expected_mbp

rank_100k %>%
    replace_na(
        list(mean_rank = 1,
            min_rank = 1,
            max_rank = 1,
            group_start = 100,
            group_end = 0,
            imprinted = F
        )
    ) %>%
    #filter(mean_rank > 0) %>%
    #drop_na() %>%
    ggplot(
    ) +
    #geom_point() +
    geom_step(
        aes(
            color=sample_1,
            group=sample_1,
            #size = same_sample,
            y=mean_rank-1, x=start,
            alpha = mean_rank > 1,
        ),
    )+ 
    geom_rect(
        data = both_rank_2,
        aes(xmin=start, xmax=end),
        group=NA,
        alpha=0.75,
        fill="green",
        ymin=0, ymax=4,
    ) +
    # add imprinted sites as orange rectangles
    geom_rect(
        data = rank_100k %>% filter(imprinted),
        aes(xmin=start, xmax=end),
        group=NA,
        alpha=0.75,
        fill="orange",
        ymin=0, ymax=4,
    ) +
    scale_x_continuous("Genomic position", label=comma) + 
    scale_y_continuous("# of inter-sample haplotypes with higher epigenetic similarity") +
    scale_color_manual("", values=c("red", "blue")) +
    scale_alpha_manual("", values=c(0.5, 0.75)) +
    #facet_zoom(xlim = c(28510120, 33480577), zoom.size = 1) +
    facet_col(~chrom) +
    theme_minimal_grid()+
    theme(legend.position="top")
#my_ggsave("Figures/MHC-different-epi-haplotypes-rank.pdf", width=12, height=8)
my_ggsave("Figures/All-different-epi-haplotypes-rank.pdf", width=12, height=22)
```

# check if we have more overlap than expected with SDs or ALTs
```{r}
merged_both_rank2 = bed_merge(both_rank_2, max_dist=1e5+1,
    is_SD = sum(is_SD)>0, is_Alt = sum(is_Alt)>0, is_imprinted = sum(imprinted)>0,
) %>% mutate(
    group="observed",
)
merged_both_rank2

merged_both_rank2 %>% 
    fwrite("Tables/both-rank-2.bed", sep="\t", quote=F, na="NA")


random_windows = merged_both_rank2 %>%
    select(chrom, start, end) %>%
    expand_grid(
        group = paste("null group x", seq(10000)),
    ) %>%
    bed_shuffle(FAI %>% filter(!name %in% c("chrM", "chrEBV", "chrX", "chrY")), seed = 9830491) %>%
    bed_map(SDs, is_SD = n() > 0) %>% 
    bed_map(ALTs, is_Alt = n() > 0) %>% 
    bed_map(imprinted, is_imprinted = n() > 0) %>%
    replace_na(list(is_SD=F, is_Alt=F, is_imprinted=F)) 
random_windows

null_both_rank2 = bind_rows(random_windows, merged_both_rank2) %>%
    group_by(group) %>%
    dplyr::summarize(
        SD = mean(is_SD), 
        Alt = mean(is_Alt)
    ) %>%
    mutate(
        observed = case_when(
            group == "observed" ~ "observed",
            TRUE ~ "10,000 random shuffles (null distribution)"
        )
    )

null_both_rank2 %>%
    pivot_longer(
        cols = c(Alt, SD),
        names_to = "type",
        values_to = "value"
    ) %>%
    ggplot(aes(x=value, fill=observed, color=observed)) +
    geom_histogram(
        data = . %>% filter(observed != "observed"),
        binwidth=0.001
    ) +
    geom_vline(
        data = . %>% filter(observed == "observed"),
        aes(xintercept=value, color=observed),
        linetype="dashed",
    ) + 
    geom_text(
        data = . %>% filter(observed == "observed"),
        aes(y=0, label=percent(value, accuracy=0.01)),
        nudge_x = 0.075,
        nudge_y = 2000,
    ) +
    #geom_density(alpha=0.5) +
    scale_fill_manual("", values=c("black", "darkorange")) +
    scale_color_manual("", values=c("black", "darkorange")) +
    scale_x_continuous(
        "Percent of windows with overlap",
        labels=percent,
    ) +
    scale_y_continuous(
        "Number of randomizations",
    ) +
    theme_minimal_grid(font_size=8) +
    facet_col(~type) +
    theme(legend.position="top")

null_both_rank2 %>%
    pivot_longer(
        cols = c(Alt, SD),
        names_to = "type",
        values_to = "value"
    ) %>%
    group_by(type) %>%
    dplyr::summarize(
        z = 1 - mean(value[observed=="observed"] > value[observed!="observed"]),
        x=value[observed=="observed"], 
    )
     

my_ggsave("Figures/Overlap-with-alts-or-SDs-for-inter-hap-ranks.pdf", width=4, height=3)
table(null_both_rank2$average_alt_overlap)
```



# get the average rank of a window for 100kbp windows per sample
```{r}
rank_100k %>%
    filter(mean_rank > 1) %>%
    ggplot(
        aes(x=mean_rank-1, weight = 1e5, fill=imprinted)
    ) +
    geom_histogram(binwidth=0.1) +
    scale_y_continuous(
        "bp",
        #trans="log10",
        #limits=c(4, NA),
        label=comma
    ) + #+ annotation_logticks(sides="l") +
    theme_cowplot() +
    facet_col(~sample_1) +
    #coord_cartesian(ylim=c(0, 100e6)) +
    scale_x_continuous(
        "mean number of inter-sample haplotypes with higher epigenetic similarity\n(per 100kbp window)",
    ) +
    scale_fill_manual(
        "Imprinted",
        values=c("black", "darkorange")
    ) +
    #facet_zoom(ylim = c(0, 100e6)) +
    theme(legend.position="top")
my_ggsave("Figures/epi-haps-mean-rank-histogram.pdf", height=6, width=8)
```







#
# test for enrichment of hap specific peaks in windows
#
```{r}
library(runner)
library(parallel)

roll_enrichment_test_df = function(d){
    tn_all_peaks = d$tn_all_peaks[1]
    tn_hap_peaks = d$tn_hap_peaks[1]
    tn_not_hap_peaks = tn_all_peaks - tn_hap_peaks
    n_all_peaks = nrow(d)
    n_hap_peaks = sum((d$p_value <= 0.05) & (abs(d$diff) > 0.25))
    n_not_hap_peaks = n_all_peaks - n_hap_peaks
    p_value = fisher.test(
        matrix(c(n_hap_peaks, n_not_hap_peaks, tn_hap_peaks, tn_not_hap_peaks),nrow=2),
        alternative="greater"
    )$p.value
    z=data.table(
        chrom = d$chrom[1],
        start = min(d$start),
        end = max(d$end),
        p_value = p_value,
        n_hap_peaks = n_hap_peaks,
        n_not_hap_peaks = n_not_hap_peaks,
        tn_hap_peaks = tn_hap_peaks,
        tn_all_peaks = tn_all_peaks,
        sample_type = d$sample_type[1]
    )
    z
}

n_in_w = 100
group_var = "sample_type"
zz = df_all %>%
    filter(chrom %in% FAI_NO_SEX$name) %>%
    dplyr::group_by(get(group_var)) %>%
    mutate(
        tn_hap_peaks = sum((p_value <= 0.05) & (abs(diff) >= 0.25)),
        tn_all_peaks = n(),
    ) %>%
    dplyr::arrange(get(group_var), chrom, start, end, .by_group = TRUE) %>%
    mutate(
        idx = seq(1, n()),
    ) %>%
    data.table
table(zz[,c("tn_hap_peaks", "tn_all_peaks")])

cl <- makeForkCluster(16)
idxs = seq(n_in_w/2, nrow(zz), n_in_w/10)
k = runner(
    zz, 
    k = n_in_w, lag = -n_in_w/2,
    f = roll_enrichment_test_df,
    at = idxs,
    #na_pad = NA,
    cl=cl,
)
stopCluster(cl)
a = data.table(t(k)) %>%
    unnest() %>%
    filter(start < end) %>%
    filter(!is.na(p_value)) %>%
    ungroup() %>%
    mutate(
        p_adjust = p.adjust(p_value, method="fdr"),
        my_group = get(group_var),
    ); a
head(a$p_value)
```

```{r}
ttz=df_all %>%
    filter(p_value <= 0.05) %>%
    ggplot(aes(x=abs(diff), color=sample_type, group=sample)) + 
    geom_vline(xintercept=0.25, linetype="dashed") +
    geom_histogram() +
    facet_col(~sample_type)+
    theme_cowplot()
my_ggsave("tmp.pdf")
```

```{r}
to_use=a
#roll_windowed_hap_peaks %>% filter(p_adjust < 0.05) %>% filter(chrom == "chr6") %>% select(chrom, group_start, group_end, p_value, p_adjust, my_group, n_hap_peaks, n_all_peaks)

y_lim = ceil(max(-log10(to_use$p_value))) 
y_by = floor(y_lim/5)
z=to_use %>%
    mutate(
        # strip chr from chrom names
        chrom = factor(gsub("chr", "", chrom), levels=1:22),
        #chrom = factor(chrom, levels=FAI_NO_SEX$chrom),
    ) %>%
    #filter(p_value < 0.05) %>%
    ggplot(
        aes(x=(
            end+start)/2,
            y=p_value,
            color=p_adjust < 0.05,
            alpha=p_adjust < 0.05,
        )
    ) +
    geom_rect(
        data = data.table(
            start=0,end=0,p_value=1,color=NULL,
            xmin=28510120, xmax=33480577, chrom=factor("6"), ymin=1, ymax=10**-y_lim
        ),
        aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax, color=NULL),
        alpha=0.25,
        fill="blue",
    ) +
    geom_hline(
        data = . %>% dplyr::group_by(my_group) %>% filter(p_adjust<0.05) %>% dplyr::summarize(y = max(p_value)),
        aes(yintercept=y),
        linetype="dashed",
        linewidth=0.1,
    )+
    geom_point(size=0.25) +
    facet_grid(my_group~chrom, space="free_x", scale="free") +
    scale_x_continuous("Genomic position (Mbp)", 
        breaks = seq(0, 300e6, 50e6),
        label=function(x) comma(x/1e6)
    ) +
    scale_color_manual("Corrected p-value < 0.05", values=c("black", "red")) +
    scale_alpha_manual("Corrected p-value < 0.05", values=c(0.05, 1)) +
    scale_y_continuous("p-value",
        trans=reverselog_trans(base=10),
        breaks=10**(-seq(0, y_lim, y_by)),
        minor_breaks=10**(-seq(0, y_lim, 0.1)),
        labels=scientific_10,
    ) +
    theme_cowplot(font_size=6) +
    theme(
        legend.position="top",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        strip.clip = "off",
        strip.background =element_rect(fill="white"),
        #strip.text.x = element_text(size = 4, color = "black")
    ) 
my_ggsave("Figures/epi-haps-binom.pdf", width=8, height=3)
```

