```{r}
source("Rscripts/utils.R")
source("Rscripts/Fire-peak-df.R")
```

```{r}
vcf="data/GM12878.vcf.gz"
vcf_df = fread(cmd=glue("bcftools view -f 'PASS' -i 'GT=\"HET\"' {vcf} | grep -v ^##"))  %>% 
    mutate(start = POS-1) %>%
    rename(
        chrom="#CHROM",
        end=POS
    )
pbsv_vcf = 
```

```{r}
set.seed(1) 
nrow(vcf_df)/1e6

hap_peaks=fire_df %>%
    mutate(
        autosome = case_when(
            chrom %in% FAI_NO_SEX$name ~ "Autosome",
            chrom == "chrX" ~ "X",
            TRUE ~ "Other"    
        ),
    ) %>%
    filter(!is.na(diff)) %>%
    filter(p_adjust<=0.05 & autosome=="Autosome" ) %>%
    data.table
sum(hap_peaks$imprinted)
rand_hap_peaks = bed_shuffle(hap_peaks, FAI)
sum(rand_hap_peaks$imprinted)

mean(hap_peaks$end-hap_peaks$start)
mean(vcf_df$end-vcf_df$start)

hap_peaks_with_null = bind_rows(
    list(real=hap_peaks, shuffle=rand_hap_peaks),
    .id="type"
    ) %>%
    #select(-start.other, end.other) %>%
    bed_map(vcf_df,
        n_variants = n(),
    ) %>% 
    replace_na(list(
        n_variants=0
    )) %>%
    mutate(
        str_imprinted = case_when(
            imprinted == TRUE ~ "imprinted",
            TRUE ~ "not-imprinted",
        )
    )
hap_peaks_with_null %>% group_by(type) %>% summarise(sum(imprinted), n())

p_sum = hap_peaks_with_null %>% group_by(type) %>%
    summarise(
        n_with_variants = sum(n_variants > 0),
        n_without_variants = sum(n_variants == 0)
    ) %>%
    ungroup %>%
    pivot_wider(names_from=type, values_from=starts_with("n_")) %>%
    rowwise %>%
    mutate(
        p_value = fisher.test(matrix(
            c(n_with_variants_real, n_without_variants_real, n_with_variants_shuffle, n_without_variants_shuffle),
            ncol=2),
            alternative="greater"
            )$p.value,        
    )
p_sum
```

```{r}
table(hap_peaks_with_null[hap_peaks_with_null$type=="real",]$str_imprinted)
hap_peaks_with_null %>%
    filter(type=="real") %>%
    ggplot(
        aes(
            x=pmin(n_variants, 10),
        )
    ) +
    geom_histogram(
        aes(fill=str_imprinted),
        binwidth=1,
        #color="black",
        width=0.4,
    ) +
    stat_bin(
        aes(
            y = after_stat(count),
            label=ifelse(..count.. > 0, ..count.., ""),
        ),
        color="black",
        vjust=1,
        binwidth=1,
        geom="text_repel",
        min.segment.length=unit(1000, "lines"),
        direction="y",
        size=1.5,
    ) +
    scale_x_continuous(
        glue("# of heterozygous SNVs per hap-selective peak\n(n={comma(nrow(hap_peaks))} autosomal peaks)"),
        breaks=seq(0,10),
        labels=c(seq(0,9), "10+")
    ) +
    scale_y_continuous("Number of peaks") +
    my_hgrid() +
    scale_fill_manual("",
        values=c("imprinted"="darkorange","not-imprinted"="darkgrey")
    ) +
    ggtitle(glue(
    "Fisher's exact one-sided p-value: {scientific(p_sum$p_value)}\n",
    "Real data: with {p_sum$n_with_variants_real} without {p_sum$n_without_variants_real};   ",
    "Shuffled data: with {p_sum$n_with_variants_shuffle} without {p_sum$n_without_variants_shuffle}"
    )) +
    theme(
        legend.position="top",
        plot.title = element_text(size = 5, face = "bold"),
    ) 

my_ggsave("Figures/n-variants-per-hap-peak.pdf", height=3, width=3)
```

```{r}
hap_peaks_with_null %>%
    filter(type=="real") %>%
    mutate(
        diff_5mC = abs(hap1_5mC_percent - hap2_5mC_percent),
        has_variant = case_when(
            n_variants > 0 ~ "With variants",
            TRUE ~ "Without variants"
        )
    ) %>%
    ggplot(
        aes(
            x=diff_5mC,
            fill=has_variant,
        )
    ) +
    geom_histogram(
        aes(y=..density..), 
        binwidth=5, alpha=0.5, position="dodge", color="black"
    ) +
    scale_fill_manual("",
        values=c("With variants"="darkblue","Without variants"="darkgrey")
    )+
    scale_x_continuous(
        "% difference in mCpG levels between haplotypes",
    ) +
    #geom_density(adjust = 0.1, alpha=0.5)+
    facet_col(~str_imprinted) +
    my_grid() +
    theme(
        legend.position="top",
    ) 
my_ggsave("Figures/5mC-difference-in-hap-selective-peaks.pdf", height=3, width=3)
```

```{r}
```