```{r}
source("Rscripts/utils.R")
source("Rscripts/Fire-peak-df.R")
```

```{r}

```

```{r}
set.seed(1) 
nrow(vcf_df)/1e6

hap_peaks=fire_df %>%
    mutate(
        autosome = case_when(
            chrom %in% FAI_NO_SEX$name ~ "Autosome",
            chrom == "chrX" ~ "X",
            TRUE ~ "Other"    
        ),
    ) %>%
    filter(!is.na(diff)) %>%
    filter(p_adjust<=0.05 & autosome=="Autosome" ) %>%
    data.table
sum(hap_peaks$imprinted)
rand_hap_peaks = bed_shuffle(hap_peaks, FAI)
sum(rand_hap_peaks$imprinted)

mean(hap_peaks$end-hap_peaks$start)
mean(vcf_df$end-vcf_df$start)

hap_peaks_with_null = bind_rows(
    list(real=hap_peaks, shuffle=rand_hap_peaks),
    .id="type"
    ) %>%
    #select(-start.other, end.other) %>%
    bed_map(vcf_df,
        n_SNVs = n(),
    ) %>% 
    bed_map(pbsv_ends,
        n_SVs = n(),
    ) %>%
    replace_na(list(
        n_SNVs=0,
        n_SVs=0
    )) %>%
    mutate(
        str_imprinted = case_when(
            imprinted == TRUE ~ "imprinted",
            TRUE ~ "not-imprinted",
        ),
        has_SVs = case_when(
            n_SVs > 0 ~ "has SV(s)",
            TRUE ~ "no SVs",
        ),
        n_variants = n_SNVs+n_SVs,
        has_variants = case_when(
            n_variants > 0 ~ "With variants",
            TRUE ~ "Without variants",
        ),
    )
hap_peaks_with_null %>% group_by(type) %>% summarise(
    sum(n_SVs),
    sum(n_SVs>0),
)
hap_peaks_with_null %>% group_by(type) %>% summarise(sum(imprinted), n())

p_sum = hap_peaks_with_null %>% group_by(type) %>%
    summarise(
        n_with_variants = sum(n_variants > 0),
        n_without_variants = sum(n_variants == 0)
    ) %>%
    ungroup %>%
    pivot_wider(names_from=type, values_from=starts_with("n_")) %>%
    rowwise %>%
    mutate(
        p_value = fisher.test(matrix(
            c(n_with_variants_real, n_without_variants_real, n_with_variants_shuffle, n_without_variants_shuffle),
            ncol=2),
            alternative="greater"
            )$p.value,        
    )
p_sum
```

```{r}
hap_peaks_with_null %>%
    filter(type=="real") %>%
    ggplot(
        aes(
            x=pmin(n_variants, 10),
        )
    ) +
    geom_histogram(
        aes(fill=str_imprinted),
        binwidth=1,
        #color="black",
        width=0.4,
    ) +
    stat_bin(
        aes(
            y = after_stat(count),
            fill=str_imprinted,
            label=ifelse(..count.. > 0, ..count.., ""),
        ),
        #color="black",
        vjust=1,
        binwidth=1,
        geom="text_repel",
        min.segment.length=unit(1000, "lines"),
        direction="y",
        size=1.5,
    ) +
    scale_x_continuous(
        glue("# of heterozygous variants (SNVs,indels,SVs)\nper hap-selective peak (n={comma(nrow(hap_peaks))} autosomal peaks)"),
        breaks=seq(0,10),
        labels=c(seq(0,9), "10+")
    ) +
    scale_y_continuous("Number of peaks") +
    my_hgrid() +
    scale_fill_manual("",
        values=c("imprinted"="darkorange","not-imprinted"="darkgrey")
    ) +
    ggtitle(glue(
    "Fisher's exact one-sided p-value: {scientific(p_sum$p_value)}\n",
    "Real data: with {p_sum$n_with_variants_real} without {p_sum$n_without_variants_real};   ",
    "Shuffled data: with {p_sum$n_with_variants_shuffle} without {p_sum$n_without_variants_shuffle}"
    )) +
    theme(
        legend.position="top",
        plot.title = element_text(size = 5, face = "bold"),
    ) 

my_ggsave("Figures/hap-selective-n-variants-per-peak.pdf", height=3, width=3)
```

```{r}
not_hap_sel_df = fire_df %>% filter(p_adjust >= 0.05 | is.na(p_adjust)) %>% mutate(
    has_variants = "Non-haplotype-selective peaks"
) %>%  mutate(
        str_imprinted = case_when(
            imprinted == TRUE ~ "imprinted",
            TRUE ~ "not-imprinted",
        ))
bw=5
asdf = bind_rows(
        list(hap_peaks_with_null %>%
            filter(type=="real"),
            not_hap_sel_df
        )
    ) %>%
    mutate(
        #diff_5mC = ave_hap_5mC_diff,
        diff_5mC = max_hap_5mC_diff,
    ) %>%
    filter(!is.na(diff_5mC)) %>%
    data.table

asdf2 = copy(asdf %>% filter(has_variants!="Non-haplotype-selective peaks"))
asdf2$has_variants = "All haplotype-selective peaks" 

bind_rows(list(asdf,asdf2)) %>%
    mutate(
        has_variants = factor(has_variants, levels=c("Non-haplotype-selective peaks", "All haplotype-selective peaks", "With variants", "Without variants")),
    ) %>%
    ggplot(
        aes(
            x=diff_5mC,
            #fill=has_variants,
            #fill=str_imprinted,
        )
    ) +
    geom_histogram(
        aes(y=bw* after_stat(density) ),
        binwidth=bw, alpha=0.5, color="black"
    ) +
    geom_text(
        data = . %>%
            dplyr::group_by(str_imprinted, has_variants) %>%
            dplyr::summarise(count = n(), stat = mean(diff_5mC)/100),
        aes(x=50, y=.2, label = glue("n={comma(count)}\nmean={percent(stat)}")),
        size=1.5,
    ) +
    #geom_density() +
    scale_fill_manual("",
        #values=c("With variants"="darkblue","Without variants"="darkgrey")
        values=c("imprinted"="darkorange","not-imprinted"="darkgrey")
    )+
    scale_x_continuous(
        "Max % difference in mCpG levels between haplotypes",
    ) +
    scale_y_continuous(
        "% of peaks",
        label=percent
    )+
    #geom_density(adjust = 0.1, alpha=0.5)+
    facet_grid(str_imprinted~has_variants, scale="free_y") +
    my_grid() +
    theme(
        legend.position="top",
    ) 
my_ggsave("Figures/hap-selective-5mC-difference-in-peaks.pdf", height=3, width=5)
```

```{r}
hap_peaks_with_null %>%
    filter(type=="real") %>%
    ggplot(
        aes(
            x=abs(diff),
        )
    ) +
    geom_histogram(
        aes(fill=str_imprinted),
        binwidth=0.01,
        #color="black",
        width=0.4,
    ) +
    scale_x_continuous(
        "Percent difference in accessibility between haplotypes",
        #limits = c(0,NA), 
        labels=percent,
    ) +
    scale_y_continuous("Number of peaks") +
    my_hgrid() +
    scale_fill_manual("", values=c("imprinted"="darkorange","not-imprinted"="darkgrey")) +
    facet_col(~has_variants, scale="free_y") + 
    theme(
        legend.position="top",
        plot.title = element_text(size = 5, face = "bold"),
    ) 
my_ggsave("Figures/hap-selective-accessibility-difference-by-variants.pdf", height=3, width=3)
```


# ideogram with hap peaks (with and without variants)
```{r}
ideo_df = hap_peaks_with_null %>%
    filter(type=="real") 



system("rm tmp.pdf")
s = 4
pdf("tmp.pdf", height=1*s, width=8*s)

kp = karyoploteR::plotKaryotype(genome="hg38", plot.type=3, chromosomes=FAI_NO_SEX$chrom, cex=1.5)

#kpAddBaseNumbers(kp)

ws=5e6
last.max = -1
data.panel = 1
for (t in c("With variants", "Without variants")){
    data = ideo_df %>% filter(has_variants==t) %>% filter(!imprinted)
    sample.regions = toGRanges(data$chrom, data$start, data$end)
    tkp = kpPlotDensity(
        kp, data=sample.regions, 
        col=transparent("darkred",0.5),
        border=transparent("darkred", 0.05),
        data.panel = data.panel,
        window.size=ws,
    )
    
    if (last.max < tkp$latest.plot$computed.values$max.density){
        last.max = tkp$latest.plot$computed.values$max.density
    }

    kpAxis(kp, 
        ymax=tkp$latest.plot$computed.values$max.density,
        cex=1.5,
        #side="left",
        numticks=2,
        data.panel = data.panel,
    )
    kpAddLabels(kp, 
        labels = t, 
        cex=1.1,
        data.panel = data.panel, 
    )
    data.panel = 2
}

dev.off()
system("cp tmp.pdf Figures/hap-specific-peaks-ideogram-with-and-without-variants.pdf")
dev.off()


```