```{r}
source("Rscripts/Fire-peak-df.R")
```


```{r}
if(F){
    gwas.catalog = fread("Revision-1/GWAS/download-my-own-catalogs/full", sep="\t", quote="") %>%
    mutate(
        chrom = paste0("chr", CHR_ID),
        start = as.numeric(CHR_POS)-1,
        end = start+1,
        rsID=as.numeric(str_replace(SNPS, "^rs", ""))
    )
} else {
    gwas.catalog = fread(
        "Revision-1/GWAS/download-my-own-catalogs/PICS2-GWAScat-2024-09-20.simple.tbl.gz", 
        sep="\t",
    ) %>% 
    # split Linked_position_hg38 on the colon and then whitespace
    mutate(
        rsID=as.numeric(str_replace(rsID, "^rs", "")),
        chrom=CHROM,
        start=as.numeric(POS)-1,
        end=start+1,
    ) %>%
    filter(
        !is.na(rsID),
        !is.na(chrom),
        !is.na(start),
        !is.na(end),
    )
}

small.gwas.catalog = gwas.catalog %>%
    select(`DISEASE/TRAIT`, rsID, MAPPED_GENE) %>%
    unique

all.gwas.sites = gwas.catalog %>%
    select(chrom, start, end) %>%
    unique
all.gwas.sites
gwas.catalog


gwas.vcf = fread(cmd='bcftools view Revision-1/GWAS/echtvar/results/gm12878.risk.vcf.gz | grep -v "^##"') %>%
    mutate(
        chrom=`#CHROM`,
        start = POS-1,
        end = POS
    ) %>% 
    mutate(
        rsID=str_extract(INFO, "rsID=[0-9]+") %>% str_replace("rsID=", "") %>% as.numeric,
        MAF=str_extract(
            INFO,
             "FREQ=1000Genomes:[0-9]+.[0-9]+"
        ) %>% str_replace("FREQ=1000Genomes:", "") %>% as.numeric
    ) 
gwas.vcf

gwas.het.vcf = fread(cmd='bcftools view Revision-1/GWAS/echtvar/results/gm12878.het.risk.vcf.gz | grep -v "^##"') %>%
    mutate(
        chrom=`#CHROM`,
        start = POS-1,
        end = POS
    ) %>% 
    mutate(
        rsID=str_extract(INFO, "rsID=[0-9]+") %>% str_replace("rsID=", "") %>% as.numeric,
        MAF=str_extract(
            INFO,
             "FREQ=1000Genomes:[0-9]+.[0-9]+"
        ) %>% str_replace("FREQ=1000Genomes:", "") %>% as.numeric
    ) %>%
    replace_na(list(MAF=0)) 
gwas.het.vcf


het.vcf.all = fread(cmd='bcftools view Revision-1/GWAS/echtvar/results/gm12878.anno.vcf.gz | grep -v "^##"') %>%
    mutate(
        chrom=`#CHROM`,
        start = POS-1,
        end = POS
    ) %>% 
    mutate(
        rsID=str_extract(INFO, "rsID=[0-9]+") %>% str_replace("rsID=", "") %>% as.numeric,
        MAF=str_extract(
            INFO,
             "FREQ=1000Genomes:[0-9]+.[0-9]+"
        ) %>% str_replace("FREQ=1000Genomes:", "") %>% as.numeric
    ) %>%
    replace_na(list(MAF=0))
het.vcf.all
table(round(100*het.vcf.all$MAF))
```

# absolute diff in hap_selective fire elements with and without variants
```{r}
tdf = fire_df %>% mutate(
    hap_selective = case_when(
        p_adjust <= 0.05 & chrom %in% FAI_NO_SEX$name ~ "hap-selective",
        TRUE ~ "not-hap-selective"
    )
) %>%
    filter(hap_selective=="hap-selective") %>%
    mutate(
        abs_diff = abs(diff)
    )
tdf %>%
    ggplot(aes(x=abs_diff, color=has_variant)) +
    geom_step(aes(y=after_stat(density)), stat="bin", bins=25, alpha=0.75) +
    #geom_density(alpha=0.5, adjust=0.75)+
    scale_x_continuous("Difference in H1 and H2 signal in haplotype selective FIRE peaks") +
    my_grid() +
    theme(
        legend.position = "top",
    )
tdf %>%
    group_by(has_variant) %>%
    summarise(
        mean_diff = mean(abs_diff),
        median_diff = median(abs_diff),
        sd=stats::sd(abs_diff),
        n=n()
    )
my_ggsave("GWAS-affect-of-variants-on-hap-selective-fire.pdf", height=3, width=3)
```


# plot the MAF of variants within HAP selective peaks
```{r}
z= fire_df %>%
    mutate(
        hap_selective = case_when(
            p_adjust <= 0.05 & chrom %in% FAI_NO_SEX$name ~ "hap-selective",
            TRUE ~ "not-hap-selective"
        )
    ) %>%
    bed_map(
        het.vcf.all,
        MAF = min(MAF),
    ) %>%
    filter(!is.na(MAF)) %>%
    select(chrom,start,end,MAF, hap_selective, diff)
table(z$hap_selective)

tdf %>%
    group_by(has_variant) %>%
    summarise(
        mean_diff = mean(abs_diff),
        median_diff = median(abs_diff),
        sd=stats::sd(abs_diff),
        n=n()
    )
my_ggsave("GWAS-affect-of-variants-on-all-fires.pdf", height=3, width=3)

z2=bed_closest(z, gwas.het.vcf,
        suffix = c("", ".y") 
    ) %>%
    mutate(
        .dist = abs(.dist)
    ) %>%
    mutate(
        overlap_gwas = .dist <= 1,
        close_gwas = .dist <= 40e3,
    ) %>%
    filter(overlap_gwas|close_gwas) %>%
    mutate(
        case = case_when(
            overlap_gwas ~ "overlaps GWAS",
            close_gwas ~ "close to GWAS",
            TRUE ~ "far from GWAS"
        )
    )
z2
table(z$hap_selective)
table(z2$.dist[z2$.dist < 0])

z2 %>%
    ggplot(
        aes(x=MAF, color = hap_selective)
    ) +
    geom_step(
        aes(
            #y=after_stat(density),
            #linetype=case
        ),
        stat="bin", bins=50,
        alpha=0.75
    ) +
    facet_col(~hap_selective+case, scales="free_y") +
    scale_x_continuous("1,000 genomes minor allele frequency of variants within FIRE peaks") +
    scale_y_continuous(
        "count",
        limits=c(0,NA),
        #label=percent,
    ) +
    scale_color_manual("", values=c("hap-selective"="darkred", "not-hap-selective"="darkblue")) +
    my_grid() +
    theme(
        legend.position = "top",
    )

my_ggsave("Figures/MAF-within-hap-selective-peaks.pdf", height=6, width=3)


z3=z2 %>%
    filter(
        hap_selective == "hap-selective"
    ) %>%
    filter(overlap_gwas|close_gwas) 
    
n_overlap_low_maf = z3 %>% filter(overlap_gwas & MAF < 0.05) %>% nrow
n_overlap_high_maf = z3 %>% filter(overlap_gwas & MAF > 0.05) %>% nrow
n_close_low_maf = z3 %>% filter(close_gwas & MAF < 0.05) %>% filter(!overlap_gwas) %>% nrow
n_close_high_maf = z3 %>% filter(close_gwas & MAF > 0.05) %>% filter(!overlap_gwas) %>% nrow
test_data = c(n_overlap_low_maf, n_overlap_high_maf, n_close_low_maf, n_close_high_maf)
test_data
fisher.test(matrix(test_data,nrow=2), alternative="less")


z4=z3 %>%
    filter(close_gwas & ! overlap_gwas) %>%
    ungroup() %>%
    mutate(
        low_MAF = MAF < 0.05,
        abs_diff = abs(diff)
    ) %>%
    group_by(
        low_MAF
    ) %>%
    filter(!is.na(diff)) 

z4 %>%   
    dplyr::summarise(
        mean_diff = mean(abs_diff),
        #median_diff = median(abs(diff)),        
        #sd=stats::sd(abs(diff)),
        n=n()
    )
high_maf = z4 %>% filter(!low_MAF) %>% pull(abs_diff)
low_maf = z4 %>% filter(low_MAF) %>% pull(abs_diff)
mean(100*high_maf)
mean(100*low_maf)
wilcox.test(high_maf, low_maf)
```

# MAF of just gwas.het
```{r}
gwas.het.vcf %>%
    ggplot(
        aes(x=MAF)
    ) +
    geom_histogram(bins=50) +
    scale_x_continuous("1,000 genomes minor allele frequency of het. GWAS variants in GM12878") +
    scale_y_continuous("count") +
    my_grid()
my_ggsave("Figures/MAF-of-all-het-GWAS-variants.pdf", height=3, width=3)
```

# get haplotype selective peaks with shuffles that have the genetic variants
```{r}
#
# RUN THE FIRST BLOCKS OF GENETIC-VARIANTS-IN-HAP-PEAKS.RMD
# TO GET THE GM12878 VCFS 
#
set.seed(1) 
hap_peaks=fire_df %>%
    mutate(
        autosome = case_when(
            chrom %in% FAI_NO_SEX$name ~ "Autosome",
            chrom == "chrX" ~ "X",
            TRUE ~ "Other"    
        ),
    ) %>%
    filter(!is.na(diff)) %>%
    filter(p_adjust<=0.05 & autosome=="Autosome" ) %>%
    data.table

rand_hap_peaks = list()
for(i in 1:10){
    rand_hap_peaks[[i]] = bed_shuffle(hap_peaks, FAI)
}
rand_hap_peaks = bind_rows(rand_hap_peaks, .id="shuffle_n")

hap_peaks_with_null = bind_rows(
    list(
        real=hap_peaks,
        random=sample_n(fire_df, nrow(hap_peaks)),
        shuffle=rand_hap_peaks,
        all_fire = fire_df
    ),
    .id="type"
    ) %>%
    bed_map(vcf_df,
        n_SNVs = n(),
    ) %>% 
    bed_map(pbsv_ends,
        n_SVs = n(),
    ) %>%
    replace_na(list(
        n_SNVs=0,
        n_SVs=0
    )) %>%
    mutate(
        str_imprinted = case_when(
            imprinted == TRUE ~ "imprinted",
            TRUE ~ "not-imprinted",
        ),
        has_SVs = case_when(
            n_SVs > 0 ~ "has SV(s)",
            TRUE ~ "no SVs",
        ),
        n_variants = n_SNVs+n_SVs,
        has_variants = case_when(
            n_variants > 0 ~ "With variants",
            TRUE ~ "Without variants",
        ),
    ) %>%
    mutate(
        group = case_when(
            type == "real" ~ "real",
            type == "random" ~ "random",
            type == "all_fire" ~ "fire",
            TRUE ~ paste0("shuffle", shuffle_n)
        ),
    ) %>%
    select(
        -score_H1, -score_H2,
        -FDR_H1, -FDR_H2,
        -FIRE_size_mean, -FIRE_size_ssd,
        -log_FDR_H1, -log_FDR_H2,
        -FIRE_end_ssd, -FIRE_start_ssd,
    ) %>%
    # move chrom start and end to the front
    relocate(chrom, start, end) %>%
    data.table 

colnames(hap_peaks_with_null)
 
# hap_peaks_with_null = bind_rows(
#     list(
#         all = hap_peaks_with_null,
#         split = hap_peaks_with_null[group=="real"] %>% mutate(
#                     group = paste0(group, " (", has_variants, ")")
#                 ) 
#     ),
#     .id = "split"
#     ) 

unique(hap_peaks_with_null$group)
hap_peaks_with_null %>% group_by(type) %>% summarise(
    sum(n_SVs),
    mean(n_SVs>0),
    sum(n_SNVs),
    mean(n_SNVs>0),
)
hap_peaks_with_null %>% filter(group=="real") %>% summarise(
    min(p_adjust),
    max(p_adjust),
)
table(hap_peaks_with_null$group)
```

# get closest with bedtools
```{r}
system("mkdir -p temp/R/")
keep_n_closest = 5
# for each group write a different bed file
hap_peaks_with_null %>%
    arrange(chrom, start, end) %>%
    select(chrom, start, end, ID, group) %>%
    group_by(group) %>%
    # group is not included in .x so adding a copy that will be written to the file
    mutate(group2=group) %>%
    group_walk(
        ~ fwrite(.x, 
            file=paste0("temp/R/", .y$group, ".bed"),
            sep="\t",
            quote=FALSE,
            row.names=FALSE,
            col.names=FALSE,
        )
    )

gwas.het.vcf %>%
    relocate("chrom", "start", "end") %>%
    arrange(chrom, start, end) %>%
    select(chrom, start, end, rsID) %>%
    #head() %>% #select(chrom, start, end, rsID) %>%
    fwrite(
        "temp/R/gwas.het.vcf",
        sep="\t",
        quote=FALSE,
        row.names=FALSE,
        col.names=FALSE,
    )

closest = fread(
    cmd=glue(
        "bedtools closest -d -t first -k {keep_n_closest} -a temp/R/gwas.het.vcf -b temp/R/*.bed"
    )
)
colnames(closest) = c(
    "chrom", "start", "end", "rsID",
    "group_number",
    "chrom.fire", "start.fire", "end.fire", "ID", "group",
    "distance"
)
closest=closest %>%
    merge(gwas.het.vcf %>% select(-ID), by=c("chrom", "start", "end", "rsID")) %>%
    merge(
        hap_peaks_with_null %>% select(-chrom, -start, -end),
        by=c("ID", "group"),
        suffixes=c("", ".fire")
    ) %>%
    select(-group_number) %>% 
    # rank the closeness of each gwas to the closest peak
    dplyr::group_by(chrom, start, end, rsID, group) %>%
    dplyr::arrange(distance, .by_group = TRUE) %>%
    dplyr::mutate(
        rank = 1:n()
    ) %>%
    ungroup() %>%
    arrange(chrom, start, end, rsID, group, rank) %>%
    filter(
        rank == 1 | group == "real"
    ) %>%
    filter(rank <= keep_n_closest) %>%
    data.table 

table(closest$rank)
length(unique(hap_peaks_with_null$group))
```



# for every gwas find the closest haplotype selective peak
```{r}
all.gwas.df = bind_rows(list(
        all = closest,
        split = closest[group=="real"] %>% mutate(
                    group = paste0(group, " (", has_variants, ")")
        )),
        .id = "split"
    ) %>%
    mutate(
        real = case_when(
            group == "real" ~ "haplotype-selective\nFIRE peaks",
            group == "real (With variants)" ~ "haplotype-selective\nFIRE peaks (with variants)",
            group == "real (Without variants)"~"haplotype-selective\nFIRE peaks (without variants)",
            group == "random" ~ "random sample of\nFIRE peaks",
            group == "fire" ~ "all FIRE peaks",
            TRUE ~ "random intervals"
        ),
        distance = abs(distance),
    ) %>%
    mutate(
        linetype = case_when(
            real == "random intervals" | grepl("real \\(", group) ~ 2,
            TRUE ~ 1
        ),
        linetype = factor(linetype, levels=c(1,2))
    ) %>%
    data.table

gwas.df = all.gwas.df %>% filter(group!="fire")

my_colors = c(
    "haplotype-selective\nFIRE peaks" = "darkred",
    "haplotype-selective\nFIRE peaks (without variants)" = alpha("darkorange", 0.4),
    "haplotype-selective\nFIRE peaks (with variants)" = alpha("brown", 0.4),
    "random sample of\nFIRE peaks" = "darkblue",
    "random intervals" = alpha("darkgray", 0.25))
gwas.df$real = factor(gwas.df$real, levels=names(my_colors))



unique(gwas.df$real)
```


# make a cumulative distribution of the distance from the gwas to the closest haplotype selective peak
```{r}
mhc = "chr6:28,510,120-33,480,577"
mhc_start = 28510120
mhc_end = 33480577
expected_count = gwas.df %>% filter(group=="real" & rank == 1) %>% nrow

gwas.df %>% group_by(group, real) %>% summarise(n())

cum.dist.gwas = gwas.df %>%
    expand_grid(without_mhc = c(TRUE, FALSE)) %>%
    filter(
        !(without_mhc == TRUE & chrom == "chr6" & start > mhc_start & start < mhc_end)
    ) %>%
    group_by(group, real, linetype, without_mhc, rank) %>%
    mutate(
        multiplier = expected_count/n()
    ) %>% 
    group_by(group, real, distance, multiplier, linetype, without_mhc, rank) %>%
    summarise(
        count = n()
    ) %>%
    ungroup() %>%
    group_by(group, real, linetype, without_mhc, rank) %>%
    arrange(distance) %>%
    mutate(
        cumcount = cumsum(count),
        cumcount_adjusted = cumsum(count*multiplier),
        fraction = cumcount_adjusted/expected_count,
    )

head(cum.dist.gwas,20)

cum.dist.gwas %>% 
    filter(real != "random intervals") %>%
    group_by(group, real, without_mhc, rank) %>%
    summarise(
        max_distance = max(distance),
        max_fraction = max(fraction),
        n(),
        unique(multiplier)
    ) %>% print(n=50)
```

# plot of the cumulative count of GWAS peaks as a function of distance
# TODO split this into two ideas, one for hap-selective with variants, and one for hap-sel without
```{r}
cum.dist.gwas %>%
    filter(rank == 1) %>%
    filter(distance < 50e3) %>%
    ggplot(
        aes(x=distance, color=real, group=group)    
    ) +
    geom_line(aes(y=fraction, linetype=linetype)) +
    scale_x_continuous(
        "Distance from GWAS variant to the closest FIRE peak",
        label=comma
    ) +
    scale_y_continuous(
        "Cumulative fraction of GWAS variants\n",
    ) +
    scale_color_manual("",values=my_colors) +
    facet_row(~without_mhc) +
    guides(color=guide_legend(nrow=3), linetype="none")+
    theme_cowplot(font_size=8)+
    theme(
        legend.position = "top",
        legend.text = element_text(size=6),
    )
my_ggsave("Figures/distance-from-GWAS-to-closest-hap-selective-cumulative.pdf", height=3, width=5)
```

```{r}
cum.dist.gwas %>%
    filter(group == "real" | group == "random") %>%
    filter(distance < 50e3) %>%
    ggplot(
        aes(x=distance, color=real, group=paste(group, rank))    
    ) +
    geom_line(aes(y=fraction, linetype=linetype, alpha=5*1/rank)) +
    scale_x_continuous(
        "Distance from GWAS variant to the closest FIRE peak",
        label=comma
    ) +
    scale_y_continuous(
        "Cumulative fraction of GWAS variants\n",
    ) +
    scale_color_manual("",values=my_colors) +
    facet_row(~without_mhc) +
    guides(color=guide_legend(nrow=3), linetype="none")+
    theme_cowplot(font_size=8)+
    theme(
        legend.position = "top",
        legend.text = element_text(size=6),
    )
my_ggsave("Figures/distance-from-GWAS-to-closest-n-hap-selective-cumulative.pdf", height=3, width=5)

cum.dist.gwas %>%
    filter(distance < 1) %>%
    filter(!without_mhc)
```

```{r}
table(cum.dist.gwas$group)
z=cum.dist.gwas %>%
    filter(rank == 1) %>%
    ungroup %>%
    filter( `group` %in% c("real", "random") ) %>%
    select(without_mhc, group, distance, fraction) %>%
    pivot_wider(
        names_from=c("group"),
        values_from=c("fraction"),
    ) %>%
    drop_na %>%
    filter(real-random > 0 ) %>%
    filter(distance < 55e3) %>%
    ggplot(aes(x=distance, y=real-random, color=without_mhc)) +
    geom_line() +
    scale_x_continuous(
        "Distance from GWAS variant to the closest FIRE peak",
        label=comma
    ) +
    scale_y_continuous(
        "Cumulative delta between haplotype-selective\nFIRE peaks and random FIRE peaks",
        limits = c(0,NA),
    ) +
    my_grid() +
    theme(
        legend.position = "top",
    )
my_ggsave("Figures/distance-from-GWAS-to-closest-hap-selective-cumulative-diff.pdf", height=3, width=3)
```


```{r}
z=gwas.df %>%
    filter(rank == 1) %>%
    #filter(split=="all") %>%
    ggplot(aes(x=distance+1, color=real, group=group)) +
    geom_step(stat="bin", bins=100, alpha=0.75)+
    scale_x_continuous(trans='log10', label=comma) +
    annotation_logticks(sides="b") +
    #scale_y_continuous(trans='log10') +
    facet_zoom(
        xy = distance < 25e3,
        horizontal = FALSE
    ) +
    scale_color_manual("",values=my_colors) +
    theme_cowplot(font_size=8)
my_ggsave("Figures/distance-from-GWAS-to-closest-hap-selective.pdf", height=3, width=5)
```

# plot the MAF of the close GWAS variants
```{r}
gwas.df %>%
    filter(real=="haplotype-selective\nFIRE peaks" | real=="random sample of\nFIRE peaks") %>%
    filter(rank == 1) %>%
    filter(
        distance < 40e3 | real == "random sample of\nFIRE peaks"
    ) %>%
    ggplot(aes(x=MAF, color=real)) +
    geom_density() +
    scale_x_continuous("1,000 genomes minor allele frequency of GWAS variants\nwith a near (< 40 kbp) FIRE peak") +
    scale_y_continuous("Density") +
    scale_color_manual("",values=my_colors) +
    theme_cowplot(font_size=8)
my_ggsave("Figures/MAF-of-GWAS-closest-to-hap-selective-peaks.pdf", height=3, width=5)
```


#
# test enrichment of GWAS in peaks
#
# test enrichment in atac peaks
#
```{r}
sum(!fire_df$is_atac_peak)
dnase_peaks

atac_only = atac_peaks %>%
    mutate(is_atac_peak=T) %>%
    bed_map(
        bed_merge(bind_rows(dnase_peaks, fire_df)),
        other_count = n()
    ) %>%
    replace_na(list(other_count=0)) %>%
    filter(other_count==0) 

dnase_only = dnase_peaks %>%
    mutate(is_dnase_peak=T) %>%
    bed_map(
        bed_merge(bind_rows(atac_peaks, fire_df)),
        other_count = n()
    ) %>%
    replace_na(list(other_count=0)) %>%
    filter(other_count==0)


fire_only = fire_df %>%
    mutate(is_fire_peak=T) %>%
    bed_map(
        bed_merge(bind_rows(atac_peaks, dnase_peaks)),
        other_count = n()
    ) %>%
    replace_na(list(other_count=0)) %>%
    filter(other_count==0)


all.gwas.validation = bind_rows(
    list(
        `All FIRE peaks` = fire_df,
        `FIRE only peaks` = fire_only,
        `FIREs that overlap ATAC/DNase` = fire_df %>%
            filter(is_atac_peak==TRUE | is_dnase_peak==TRUE),
        `Random windows` = bed_shuffle(fire_df, FAI, excl=fire_df) %>% mutate(is_atac_peak=TRUE),
        `ATAC only peaks` = atac_only %>% filter(is_atac_peak==TRUE),
        `All ATAC peaks` = atac_peaks %>% mutate(is_atac_peak=TRUE),
        `DNase only peaks` = dnase_only %>% filter(is_dnase_peak==TRUE),
        `All DNase peaks` = dnase_peaks %>% mutate(is_dnase_peak=TRUE)
    ),
    .id="case"
    ) %>%
    select(case, chrom, start, end) %>%
    bed_map(
        gwas.vcf,
        gwas_count = n()
    ) %>%
    replace_na(list(gwas_count=0, all_gwas_count=0)) %>%
    mutate(
        per_peak_gwas_per_bp = gwas_count/(end-start),
    )

gwas.validation = all.gwas.validation %>%
    group_by(case) %>%
    summarise(
        n_peaks = n(),
        gwas_per_kbp_of_peak = 1000 * sum(gwas_count)/sum(end-start),
        peaks_with_gwas = sum(gwas_count>0),
        gwas_in_peaks = sum(gwas_count),
        #all_gwas_per_kbp_of_peak = 1000 * sum(all_gwas_count)/sum(end-start),
        #peaks_with_all_gwas = sum(all_gwas_count>0),
        #all_gwas_in_peaks = sum(all_gwas_count),
        bp_of_peaks = sum(end-start),
        #frac_encode = mean(encode_count > 0),
    ) %>%
    mutate(
        bp_of_peaks_without_gwas = bp_of_peaks - gwas_in_peaks,
    ) %>%
    data.table
gwas.validation
```

```{r}
long.gwas.validation = gwas.validation %>%
    mutate(
        is_shuffled = case_when(
            case == "Random windows" ~ "_shuffle",
            TRUE ~ "",
        ),
        peaks_without_gwas = n_peaks-gwas_in_peaks,
    )
long.gwas.validation

wide_cols = c("case", "n_peaks", "peaks_with_gwas", "gwas_in_peaks",
        "gwas_per_kbp_of_peak", "peaks_without_gwas", "bp_of_peaks", "bp_of_peaks_without_gwas")
after_wide_cols = paste0(wide_cols, "_shuffle")
after_wide_cols_2 = paste0(wide_cols, "")
wide.with.shuffle.gwas.validation = long.gwas.validation %>%
    #filter(gwas_type != "all_gwas_per_kbp_of_fire_peak") %>%
    #select(-is_atac_peak, -shuffled, -gwas_type) %>%
    pivot_wider(
        names_from = c(is_shuffled),
        values_from = wide_cols,
        values_fn = list,
        names_sep = ""
    ) %>%
    unnest(after_wide_cols) %>%
    unnest(after_wide_cols_2) %>%
    rowwise() %>%
    #filter(n_peaks > 0) %>%
    mutate(
        test = (fisher.test(matrix(c(
            peaks_with_gwas,
            peaks_without_gwas,
            peaks_with_gwas_shuffle,
            peaks_without_gwas_shuffle
        ), nrow=2))$p.value),
    ) %>%
    ungroup %>% 
    mutate(
        enrichment = gwas_per_kbp_of_peak/gwas_per_kbp_of_peak_shuffle,
    ) %>%
    select(-case_shuffle) %>%
    # move enrichment to the front
    relocate(enrichment, test, .after=case) %>%
    arrange(enrichment) %>%
    data.table
wide.with.shuffle.gwas.validation

keep_cases = c("FIRE only peaks",
        "FIREs that overlap ATAC/DNase",
        "DNase only peaks", 
        "ATAC only peaks", "Random windows")
long.gwas.validation %>%
    filter(
        case %in% keep_cases
    ) %>%
    ggplot(aes(y=case, x=gwas_per_kbp_of_peak, fill=case)) +
    geom_bar(stat="identity", position="dodge", width=0.6) +
    geom_text(
        aes(
            label=paste0(round(gwas_per_kbp_of_peak, 4), "\nn=", peaks_with_gwas)
        ),
        position=position_dodge(width=0.6),
        vjust=0.5,
        size=1, 
        hjust=-0.1,
    )+
    scale_x_continuous("GWAS variants per 1,000 bp of peak")+
    ylab("")+
    #facet_col(~gwas_type, scales="free_x")+
    scale_fill_manual("", 
        values=c(
            "FIRE only peaks"="red",
            "FIREs that overlap ATAC/DNase"="darkred",
            "DNase only peaks"="darkblue",
            "ATAC only peaks"="darkblue",
            "Random windows"="darkgray"
            #NULL="darkgray"
        )
    )+
    # turn off clipping
    coord_cartesian(clip="off")+
    # drop the fill guide
    guides(fill="none")+
    my_grid()
my_ggsave("Figures/gwas-per-kbp-of-fire-peak.pdf", height=2, width=4)
```

# Test for enrichment of fire only peaks vs atac only or dnase only peaks
```{r}
fo = wide.with.shuffle.gwas.validation %>%
    filter(case == "FIRE only peaks") 
ao = wide.with.shuffle.gwas.validation %>%
    filter(case == "ATAC only peaks")
do = wide.with.shuffle.gwas.validation %>%
    filter(case == "DNase only peaks")


fisher.test(
    matrix(c(
        fo$gwas_in_peaks,
        fo$bp_of_peaks_without_gwas,
        ao$gwas_in_peaks,
        ao$bp_of_peaks_without_gwas
    ), nrow=2),
    alternative="greater"
)

fisher.test(
    matrix(c(
        fo$gwas_in_peaks,
        fo$bp_of_peaks_without_gwas,
        do$gwas_in_peaks,
        do$bp_of_peaks_without_gwas
    ), nrow=2),
    alternative="greater"
)


wide.with.shuffle.gwas.validation %>%
    group_by(case) %>%
    summarise(
        1000*sum(gwas_in_peaks)/sum(bp_of_peaks_without_gwas),
    )

gwas.validation
t.test()
```

#
# enrichment of GWAS plots
#
```{r}
z=wide.with.shuffle.gwas.validation %>%
    filter(
        case %in% keep_cases
    ) %>%
    ggplot(
        aes(y=case,
            x=enrichment, 
            fill=case
        )
    ) + 
    geom_bar(stat="identity", position="dodge", width=0.6) +
    geom_text(
        aes(
            label=paste0( "n=", peaks_with_gwas)
        ),
        position=position_dodge(width=0.6),
        vjust=0.5,
        size=1, 
        hjust=-0.1,
    )+
    scale_x_continuous(
        "Per base enrichment of GWAS variants within peaks over random windows",
    )+
    ylab("")+
    scale_fill_manual("", 
        values=c(
            "FIRE only peaks"="red",
            "FIREs that overlap ATAC/DNase"="darkred",
            "DNase only peaks"="darkblue",
            "ATAC only peaks"="darkblue",
            "Random windows"="darkgray"
            #NULL="darkgray"
        )
    )+
    # turn off clipping
    coord_cartesian(clip="off")+
    # drop the fill guide
    guides(fill="none")+
    my_grid()
my_ggsave("Figures/gwas-per-kbp-of-fire-peak-enrichment.pdf", height=2, width=5)

wide.with.shuffle.gwas.validation %>%
    fwrite("Figures/gwas-enrichment.tsv",
        sep="\t",
        quote=FALSE,
        row.names=FALSE,
        na="NA"
    )
wide.with.shuffle.gwas.validation
```


#
#
#
```{r}
colnames(gwas.df)
to_investigate = all.gwas.df %>% 
    filter(real == "haplotype-selective\nFIRE peaks" | real == "all FIRE peaks") %>%
    #filter(group=="real") %>%
    filter(distance < 40e3) %>%
    select(1:10, rsID, distance, real, start.fire, end.fire, fire_coverage.fire, coverage.fire, FDR.fire, fire_coverage_H1.fire, coverage_H1.fire, fire_coverage_H2.fire, coverage_H2.fire, p_adjust.fire, diff.fire,) %>%
    merge(small.gwas.catalog, by="rsID", all.x=TRUE) %>%
    arrange(`#CHROM`, POS, rsID)
    
table(to_investigate$real)

to_investigate %>%
    filter(real == "haplotype-selective\nFIRE peaks") %>%
    fwrite(
        "revision-1/GWAS/hap-selective-peaks/results/results-to-investigate.tsv",
        sep="\t",
        quote=FALSE,
        row.names=FALSE,
        na="NA"
    )

diseases = sort(table(to_investigate %>% filter("haplotype-selective\nFIRE peaks"==real) %>% select(`DISEASE/TRAIT`)))
diseases[diseases> 10]
length(diseases)
length(diseases[diseases> 1])
length(diseases[diseases> 10])
```

```{r}
cum.dist.gwas %>%
    filter(distance == 0) %>%
    filter(!grepl("shuffle", group)) %>%
    select(real, distance, cumcount, cumcount_adjusted)
```


# TODO test for enrichment of terms in this list vs the whole catalog 
# the null will be:
# any het gwas that is within ~20kbp of any FIRE peak
```{r}
disease_with_enrichment = to_investigate %>%
    group_by(real, `DISEASE/TRAIT`) %>%
    summarise(count = n()) %>%
    filter(count > 1) %>%
    pivot_wider(names_from=real, values_from=count) %>%
    rename(
        hap_count = `haplotype-selective\nFIRE peaks`,
        all_count = `all FIRE peaks`
    ) %>%
    mutate(
        all_count_total = sum(all_count, na.rm=TRUE),
        hap_count_total = sum(hap_count, na.rm=TRUE),
    ) %>%
    drop_na %>%
    rowwise %>%
    mutate(
        hap_per = 100*hap_count/hap_count_total,
        all_per = 100*all_count/all_count_total,
        p_value = fisher.test(matrix(c(hap_count, hap_count_total-hap_count, all_count, all_count_total-all_count), nrow=2), 
            #alternative = "greater"
        )$p.value
    ) %>%
    ungroup() %>%
    mutate(
        p_adjust = p.adjust(p_value, method="BH"),
        log2fc = log2(hap_per) - log2(all_per) 
    ) %>%
    data.table 

disease_with_enrichment %>%
    filter(p_value < 0.05)  %>%
    filter(p_adjust < 0.05) %>%
    #filter(hap_count > 2) %>%
    arrange(p_value)

```


# make a volcano plot of the enrichment of diseases in the hap-selective peaks
```{r}
min_adjust_p_value = disease_with_enrichment %>% filter(p_adjust < 0.05) %>% pull(p_value) %>% max 
xlims = disease_with_enrichment %>% pull(log2fc) %>% range
xlims[1] = floor(xlims[1])
xlims[2] = ceiling(xlims[2])
xlims
disease_with_enrichment %>%
    ggplot(
        aes(
            x=log2fc,
            y=-log10(p_value), 
            color = (p_adjust < 0.05) 
        )
    ) +
    geom_vline(xintercept=c(-2,2), linetype="dashed", color="darkblue") +
    geom_segment(
        y=-log10(min_adjust_p_value),
        yend=-log10(min_adjust_p_value),
        x=xlims[1]-0.5,
        xend=xlims[2]+0.5,
        linetype="dashed", color="darkblue"
    ) +
    geom_point(
        aes(
            #size=hap_count
        ),
        size=0.5,
    ) +
    geom_text_repel(
        data = . %>% filter(p_adjust<0.05 & abs(log2fc) > 2),
        aes(label=`DISEASE/TRAIT`),
        size=1.5,
        min.segment.length=0,
        segment.color=alpha("black", 0.75),
        segment.size=0.25,
        color="black",
        max.overlaps=Inf,
        direction="y",
        #nudge_y=10,
        force_pull=0.1,
        force=10,
        seed=42,
        xlim  = c(xlims[2]+1, NA), hjust=0,
        #xlim  = c(10, NA), hjust=1,
    ) + 
    scale_x_continuous(
        "log2 fold change in enrichment for disease association",
        limits=c(NA,15),
        breaks=seq(xlims[1], xlims[2], 1)
    ) +
    scale_y_continuous(
        "-log10(p-value)",
    ) +
    scale_size_area("count", trans="log10") +
    scale_color_manual(
        "Significant", 
        values=c("TRUE"=alpha("darkred", 0.6), "FALSE"=alpha("gray", 0.2))
    ) +
    theme_cowplot(
        font_size=8,
        #base_size=6
    ) +
    theme(
        legend.position = "none",
    )
my_ggsave("Figures/disease-enrichment-volcano.pdf", height=2, width=4)
```

# make a table for Immunoglobulin A vasculitis
```{r}
to_investigate %>%
    filter(real == "haplotype-selective\nFIRE peaks") %>%
    filter(`DISEASE/TRAIT` == "Immunoglobulin A vasculitis") %>%
    mutate(
        start = POS-1,
        end = POS,
    ) %>%
    relocate(
        `#CHROM`, start, end, rsID, REF, ALT, 
    ) %>% 
    select(
        -POS,
        -split, 
        -real,
        -FORMAT,
        -INFO,
    ) %>%
    arrange(`#CHROM`, start, end) %>%
    fwrite(
        "revision-1/GWAS/hap-selective-peaks/results/Immunoglobulin-A-vasculitis.bed",
        sep="\t",
        quote=FALSE,
        row.names=FALSE,
        na="NA"
    )
```


# make tables to share of hap peaks
```{r}
table(hap_peaks_with_null$type)

hap_peaks_with_null %>%
    filter(type=="all_fire") %>%
    select(-type) %>%
    arrange(chrom,start,end) %>%
    rename(
        `#chrom` = chrom,
    ) %>%
    fwrite(
        file="revision-1/GWAS/hap-selective-peaks/results/annotated-fire-peaks.bed.gz",
        sep="\t",
        quote=FALSE,
        row.names=FALSE,
        na="NA"
    )

hap_peaks_with_null %>%
    filter(type=="real") %>%
    select(-type) %>%
    arrange(chrom,start,end) %>%
    rename(
        `#chrom` = chrom,
    ) %>%
    fwrite(
        file="revision-1/GWAS/hap-selective-peaks/results/annotated-hap-selective-peaks.bed.gz",
        sep="\t",
        quote=FALSE,
        row.names=FALSE,
        na="NA"
    )

hap_peaks_with_null %>%
    filter(type=="real") %>%
    filter(has_variants == "With variants") %>%
    select(-type) %>%
    arrange(chrom,start,end) %>%
    rename(
        `#chrom` = chrom,
    ) %>%
    fwrite(
        file="revision-1/GWAS/hap-selective-peaks/results/annotated-hap-selective-peaks-with-variants.bed.gz",
        sep="\t",
        quote=FALSE,
        row.names=FALSE,
        na="NA"
    )
```


```{r}
fire_df %>%
    filter(chrom !="chrX" & chrom !="chrY") %>%
    filter(!is.na(p_adjust)) %>%
    group_by(p_adjust <= 0.05,  has_variant) %>%
    summarise(
        n_with_ctcf_peak = sum(is_ctcf_peak),
        count=n(),
        percent_with_ctcf_peak = 100*n_with_ctcf_peak/n(),
    )
#
#TRUE               FALSE                    110    554                   19.9
#TRUE               TRUE                     158    677                   23.3
fisher.test(matrix(
    c(110, 554, 158, 677), nrow=2
    ),
    alternative="less"
)
```
